
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.3.1, mkdocs-material-8.5.4">
    
    
      
        <title>Bedmethyl - EpiJinn</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.80dcb947.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.cbb835fc.min.css">
        
          
          
          <meta name="theme-color" content="#4cae4f">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="green" data-md-color-accent="lightgreen">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#module-epijinnbedmethyl" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="EpiJinn" class="md-header__button md-logo" aria-label="EpiJinn" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            EpiJinn
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Bedmethyl
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    EpiJinn
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="EpiJinn" class="md-nav__button md-logo" aria-label="EpiJinn" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    EpiJinn
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    EpiJinn
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2_1" type="checkbox" id="__nav_2_1" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2_1">
          Epijinn
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Epijinn" data-md-level="2">
        <label class="md-nav__title" for="__nav_2_1">
          <span class="md-nav__icon md-icon"></span>
          Epijinn
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Bedmethyl
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Bedmethyl
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functions" class="md-nav__link">
    Functions
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read_sample_sheet" class="md-nav__link">
    read_sample_sheet
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_columns" class="md-nav__link">
    subset_bed_columns
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bedresult" class="md-nav__link">
    BedResult
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bedmethylitem" class="md-nav__link">
    BedmethylItem
  </a>
  
    <nav class="md-nav" aria-label="BedmethylItem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binarize_bed" class="md-nav__link">
    binarize_bed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_mod_subtype" class="md-nav__link">
    subset_bed_to_mod_subtype
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotate_record" class="md-nav__link">
    annotate_record
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_analysis" class="md-nav__link">
    perform_analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_record" class="md-nav__link">
    plot_record
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_base_matches" class="md-nav__link">
    subset_bed_to_base_matches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_pattern_match" class="md-nav__link">
    subset_bed_to_pattern_match
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bedmethylitemgroup" class="md-nav__link">
    BedmethylItemGroup
  </a>
  
    <nav class="md-nav" aria-label="BedmethylItemGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods_1" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_analysis_in_bedmethylitemgroup" class="md-nav__link">
    perform_all_analysis_in_bedmethylitemgroup
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#customtranslator" class="md-nav__link">
    CustomTranslator
  </a>
  
    <nav class="md-nav" aria-label="CustomTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_1" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_2" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#patterntranslator" class="md-nav__link">
    PatternTranslator
  </a>
  
    <nav class="md-nav" aria-label="PatternTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro_1" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables_1" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_2" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot_1" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_3" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color_1" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth_1" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color_1" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict_1" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html_1" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_1" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color_1" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text_1" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth_1" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features_1" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot_1" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature_1" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record_1" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Methyl/" class="md-nav__link">
        Methyl
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../dnd/" class="md-nav__link">
        Dnd
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reports/" class="md-nav__link">
        Reports
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../version/" class="md-nav__link">
        Version
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functions" class="md-nav__link">
    Functions
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read_sample_sheet" class="md-nav__link">
    read_sample_sheet
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_columns" class="md-nav__link">
    subset_bed_columns
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bedresult" class="md-nav__link">
    BedResult
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bedmethylitem" class="md-nav__link">
    BedmethylItem
  </a>
  
    <nav class="md-nav" aria-label="BedmethylItem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binarize_bed" class="md-nav__link">
    binarize_bed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_mod_subtype" class="md-nav__link">
    subset_bed_to_mod_subtype
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotate_record" class="md-nav__link">
    annotate_record
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_analysis" class="md-nav__link">
    perform_analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_record" class="md-nav__link">
    plot_record
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_base_matches" class="md-nav__link">
    subset_bed_to_base_matches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_bed_to_pattern_match" class="md-nav__link">
    subset_bed_to_pattern_match
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bedmethylitemgroup" class="md-nav__link">
    BedmethylItemGroup
  </a>
  
    <nav class="md-nav" aria-label="BedmethylItemGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods_1" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_analysis_in_bedmethylitemgroup" class="md-nav__link">
    perform_all_analysis_in_bedmethylitemgroup
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#customtranslator" class="md-nav__link">
    CustomTranslator
  </a>
  
    <nav class="md-nav" aria-label="CustomTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_1" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_2" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#patterntranslator" class="md-nav__link">
    PatternTranslator
  </a>
  
    <nav class="md-nav" aria-label="PatternTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro_1" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables_1" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_2" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot_1" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_3" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color_1" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth_1" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color_1" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict_1" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html_1" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_1" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color_1" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text_1" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth_1" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features_1" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot_1" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature_1" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record_1" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn/edit/main/reference/epijinn/Bedmethyl.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


<h1 id="module-epijinnbedmethyl">Module epijinn.Bedmethyl</h1>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="c1"># Copyright 2024 Edinburgh Genome Foundry, University of Edinburgh</span>

<span class="c1">#</span>

<span class="c1"># This file is part of EpiJinn.</span>

<span class="c1">#</span>

<span class="c1"># EpiJinn is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.</span>

<span class="c1">#</span>

<span class="c1"># EpiJinn is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</span>

<span class="c1">#</span>

<span class="c1"># You should have received a copy of the GNU General Public License along with Ediacara. If not, see &lt;https://www.gnu.org/licenses/&gt;.</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">copy</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">Bio</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">dna_features_viewer</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.Methyl</span><span class="w"> </span><span class="kn">import</span><span class="w"> </span><span class="n">METHYLASES</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.Methyl</span><span class="w"> </span><span class="kn">import</span><span class="w"> </span><span class="n">annotate_methylation</span>

<span class="n">COMPLEMENTS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>

<span class="w">    </span><span class="s2">&quot;A&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;T&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;C&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;G&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;G&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;C&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;T&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;A&quot;</span><span class="p">,</span>

<span class="p">}</span>

<span class="c1"># From https://github.com/nanoporetech/modkit/blob/master/book/src/intro_bedmethyl.md</span>

<span class="n">BEDMETHYL_HEADER</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span>

<span class="w">    </span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;start_position&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;end_position&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;modified_base_code_and_motif&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;score&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;strand&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;strand_start_position&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;strand_end_position&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;color&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Nvalid_cov&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;percent_modified&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Nmod&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Ncanonical&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Nother_mod&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Ndelete&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Nfail&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Ndiff&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="s2">&quot;Nnocall&quot;</span><span class="p">,</span>

<span class="p">]</span>

<span class="c1"># Remove duplicate and unnecessary columns from report:</span>

<span class="n">columns_for_pdf_report</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span>

<span class="w">    </span><span class="n">BEDMETHYL_HEADER</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">i</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="mi">5</span><span class="p">,</span><span class="w"> </span><span class="mi">9</span><span class="p">,</span><span class="w"> </span><span class="mi">10</span><span class="p">,</span><span class="w"> </span><span class="mi">11</span><span class="p">,</span><span class="w"> </span><span class="mi">12</span><span class="p">,</span><span class="w"> </span><span class="mi">13</span><span class="p">,</span><span class="w"> </span><span class="mi">14</span><span class="p">,</span><span class="w"> </span><span class="mi">15</span><span class="p">,</span><span class="w"> </span><span class="mi">16</span><span class="p">,</span><span class="w"> </span><span class="mi">17</span><span class="p">]</span>

<span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="p">[</span>

<span class="w">    </span><span class="s2">&quot;status&quot;</span>

<span class="p">]</span><span class="w">  </span><span class="c1"># added during binarization</span>

<span class="n">DATA_DIR</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)),</span><span class="w"> </span><span class="s2">&quot;data&quot;</span><span class="p">)</span>

<span class="c1"># For looking up modified_base_code_and_motif entries in bedmethyl:</span>

<span class="n">MODIFICATION_CODES</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;mod_base_codes.csv&quot;</span><span class="p">))</span>

<span class="c1"># Adapted From https://github.com/samtools/hts-specs/blob/master/SAMtags.pdf</span>

<span class="k">class</span><span class="w"> </span><span class="nc">CustomTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="o">.</span><span class="n">BiopythonTranslator</span><span class="p">):</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Custom translator used for the main plot.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">compute_filtered_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Display only selected features.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">8</span><span class="w">  </span><span class="c1"># a good number of features to display</span>

<span class="w">        </span><span class="c1"># Keep longest features</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">n</span><span class="p">:</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">features</span><span class="p">:</span>

<span class="w">                </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="p">)]</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="w">            </span><span class="n">max_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature_lengths</span><span class="p">[</span><span class="n">n</span><span class="p">]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">features</span><span class="p">:</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">max_length</span><span class="p">:</span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="n">feature</span><span class="p">]</span>

<span class="w">        </span><span class="k">else</span><span class="p">:</span><span class="w">  </span><span class="c1"># no need to do anything if not enough features</span>

<span class="w">            </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">filtered_features</span>

<span class="k">class</span><span class="w"> </span><span class="nc">PatternTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="o">.</span><span class="n">BiopythonTranslator</span><span class="p">):</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Custom translator used for the pattern annotation plot.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">compute_feature_color</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">):</span>

<span class="w">        </span><span class="n">status</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;status&quot;</span>

<span class="w">        </span><span class="c1"># Values are from binarize_bed() :</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">status</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;1&quot;</span><span class="p">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="s2">&quot;red&quot;</span>

<span class="w">        </span><span class="k">elif</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">status</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;U&quot;</span><span class="p">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="s2">&quot;yellow&quot;</span>

<span class="w">        </span><span class="k">elif</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">status</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;0&quot;</span><span class="p">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="s2">&quot;grey&quot;</span>

<span class="w">        </span><span class="c1"># to be implemented later:</span>

<span class="w">        </span><span class="k">elif</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">status</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;?&quot;</span><span class="p">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="s2">&quot;tab:cyan&quot;</span>

<span class="w">        </span><span class="c1"># There should be no other options.</span>

<span class="k">class</span><span class="w"> </span><span class="nc">BedResult</span><span class="p">:</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Results of a bedmethyl table analysis.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">modification</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">):</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">feature_cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">50</span><span class="w">  </span><span class="c1"># do not create a plot if there are more features</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">modification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modification</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">mod_abbreviation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>

<span class="w">            </span><span class="n">MODIFICATION_CODES</span><span class="p">[</span><span class="s2">&quot;Code&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">modification</span>

<span class="w">        </span><span class="p">]</span><span class="o">.</span><span class="n">Abbreviation</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span>

<span class="w">            </span><span class="mi">0</span>

<span class="w">        </span><span class="p">]</span><span class="w">  </span><span class="c1"># there is only one entry so we take the first</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">mod_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>

<span class="w">            </span><span class="n">MODIFICATION_CODES</span><span class="p">[</span><span class="s2">&quot;Code&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">modification</span>

<span class="w">        </span><span class="p">]</span><span class="o">.</span><span class="n">Name</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">mod_chebi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>

<span class="w">            </span><span class="n">MODIFICATION_CODES</span><span class="p">[</span><span class="s2">&quot;Code&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">modification</span>

<span class="w">        </span><span class="p">]</span><span class="o">.</span><span class="n">ChEBI</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">unmodified_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>

<span class="w">            </span><span class="n">MODIFICATION_CODES</span><span class="p">[</span><span class="s2">&quot;Code&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">modification</span>

<span class="w">        </span><span class="p">]</span><span class="o">.</span><span class="n">Unmodified_base</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">record</span>

<span class="w">        </span><span class="c1"># FILTER (ANNOTATED) RECORD AND ADD STATUS:</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="n">labelstart</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span>

<span class="w">            </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">unmodified_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=&quot;</span>

<span class="w">        </span><span class="p">)</span><span class="w">  </span><span class="c1"># unfinished to account for +/- strands</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;@epijinn&quot;</span><span class="p">:</span><span class="w">  </span><span class="c1"># as annotated by the function</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">labelstart</span><span class="p">):</span>

<span class="w">                    </span><span class="c1"># To display seq in report:</span>

<span class="w">                    </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">unmodified_base</span>

<span class="w">                    </span><span class="c1"># Assign status:</span>

<span class="w">                    </span><span class="c1"># no need to check for strand as the complement won&#39;t be modified</span>

<span class="w">                    </span><span class="n">location_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>

<span class="w">                    </span><span class="c1"># expect exactly one bed table entry per modified nucleotide:</span>

<span class="w">                    </span><span class="n">status_symbol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>

<span class="w">                        </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">location_start</span>

<span class="w">                    </span><span class="p">]</span><span class="o">.</span><span class="n">STATUS</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="w">                    </span><span class="c1"># used by a custom BiopythonTranslator to colour the annotation:</span>

<span class="w">                    </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">status_symbol</span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="n">feature</span><span class="p">]</span>

<span class="w">        </span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_features</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">feature_cutoff</span><span class="p">:</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">img_created</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">False</span>

<span class="w">        </span><span class="k">else</span><span class="p">:</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">img_created</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">True</span>

<span class="w">            </span><span class="n">fig</span><span class="p">,</span><span class="w"> </span><span class="n">ax1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">))</span>

<span class="w">            </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PatternTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">            </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">with_ruler</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="w"> </span><span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fig</span>

<span class="k">class</span><span class="w"> </span><span class="nc">BedmethylItem</span><span class="p">:</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Analyse a bedmethyl file.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **sample**</span>

<span class="s2">    &gt; The name of the sample, for example a barcode id (`str`).</span>

<span class="s2">    **reference**</span>

<span class="s2">    &gt; A Biopython SeqRecord.</span>

<span class="s2">    **bedmethyl**</span>

<span class="s2">    &gt; A pandas dataframe of a bedmethyl file for the reference.</span>

<span class="s2">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="p">,</span><span class="w"> </span><span class="n">reference</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl</span><span class="p">):</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">sample</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span><span class="o">.</span><span class="n">name</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span><span class="o">.</span><span class="n">id</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bedmethyl</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">reference_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">perform_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Perform analysis and plot the sequence.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">methylase_str</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;methylases&quot;</span><span class="p">]</span>

<span class="w">        </span><span class="n">methylase</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">METHYLASES</span><span class="p">[</span><span class="n">methylase_str</span><span class="p">]</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">plot_record</span><span class="p">()</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span><span class="w">  </span><span class="c1"># BedResult instances. For easy reference in pug template.</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">modification</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">modified_base_code_and_motif</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>

<span class="w">            </span><span class="c1"># RUN BED SUBSET ETC</span>

<span class="w">            </span><span class="n">bed_subtype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">mod</span><span class="o">=</span><span class="n">modification</span><span class="p">)</span>

<span class="w">            </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">subset_bed_to_pattern_match</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">bed_subtype</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">bed_binarized</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">binarize_bed</span><span class="p">(</span>

<span class="w">                </span><span class="n">bed_pattern_match</span><span class="p">,</span>

<span class="w">                </span><span class="n">met_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;methylated_cutoff&quot;</span><span class="p">],</span>

<span class="w">                </span><span class="n">nonmet_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;unmethylated_cutoff&quot;</span><span class="p">],</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">final_bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset_bed_columns</span><span class="p">(</span><span class="n">bed_binarized</span><span class="p">)</span>

<span class="w">            </span><span class="n">bedresult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BedResult</span><span class="p">(</span><span class="n">modification</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">final_bed</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="o">=</span><span class="n">annotated_record</span><span class="p">)</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="n">bedresult</span><span class="p">]</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">annotate_record</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">record</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">plot_record</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="n">fig</span><span class="p">,</span><span class="w"> </span><span class="n">ax1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">))</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">CustomTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">with_ruler</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="w"> </span><span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">fig</span>

<span class="w">    </span><span class="nd">@staticmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">mod</span><span class="p">):</span>

<span class="w">        </span><span class="c1"># Columnname from Bedmethyl file specification</span>

<span class="w">        </span><span class="n">bed_subtype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;modified_base_code_and_motif&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">mod</span><span class="p">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed_subtype</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">subset_bed_to_base_matches</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="w"> </span><span class="n">modified_base</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">):</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="kc">None</span><span class="p">:</span><span class="w">  </span><span class="c1"># optional bed allows linking multiple bed subset methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modified_base</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">mod_base_complement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">COMPLEMENTS</span><span class="p">[</span><span class="n">mod_base</span><span class="p">]</span>

<span class="w">        </span><span class="c1"># POSITIVE STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_positive_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span>

<span class="w">            </span><span class="n">pos</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nb">enumerate</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">mod_base</span>

<span class="w">        </span><span class="p">]</span>

<span class="w">        </span><span class="c1"># Columnname from Bedmethyl file specification:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_positive_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># NEGATIVE STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_negative_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span>

<span class="w">            </span><span class="n">pos</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nb">enumerate</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">mod_base_complement</span>

<span class="w">        </span><span class="p">]</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_negative_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_basematch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">negative_strand_filter</span><span class="p">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed_basematch</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">subset_bed_to_pattern_match</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="kc">None</span><span class="p">:</span><span class="w">  </span><span class="c1"># optional bed allows linking multiple bed subset methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">methylated_index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">methylated_index</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">record_copy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">annotated_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotate_methylation</span><span class="p">(</span><span class="n">record_copy</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=</span><span class="p">[</span><span class="n">methylase</span><span class="p">])</span>

<span class="w">        </span><span class="c1"># CREATE LIST OF POSITIONS</span>

<span class="w">        </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="n">label_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=1)&quot;</span>

<span class="w">        </span><span class="n">label_neg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=-1)&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">annotated_record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;@epijinn&quot;</span><span class="p">:</span><span class="w">  </span><span class="c1"># as annotated by function above</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">label_pos</span><span class="p">:</span>

<span class="w">                    </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">]</span>

<span class="w">                </span><span class="k">elif</span><span class="w"> </span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">label_neg</span><span class="p">:</span>

<span class="w">                    </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">]</span>

<span class="w">        </span><span class="c1"># SUBSET USING POSITIONS</span>

<span class="w">        </span><span class="c1"># Columnname from Bedmethyl file specification:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">positive_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">negative_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">negative_strand_filter</span><span class="p">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span>

<span class="w">    </span><span class="nd">@staticmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">binarize_bed</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">met_cutoff</span><span class="p">,</span><span class="w"> </span><span class="n">nonmet_cutoff</span><span class="p">):</span>

<span class="w">        </span><span class="n">bed</span><span class="p">[</span><span class="s2">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;U&quot;</span><span class="w">  </span><span class="c1"># prefill undetermined</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">index</span><span class="p">,</span><span class="w"> </span><span class="n">row</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">bed</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;percent_modified&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="nb">float</span><span class="p">(</span><span class="n">met_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">:</span>

<span class="w">                </span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">index</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1&quot;</span><span class="w">  </span><span class="c1"># methylated, symbol may change</span>

<span class="w">            </span><span class="k">elif</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;percent_modified&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nb">float</span><span class="p">(</span><span class="n">nonmet_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">:</span>

<span class="w">                </span><span class="n">bed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">index</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;0&quot;</span><span class="w">  </span><span class="c1"># unmethylated</span>

<span class="w">            </span><span class="c1"># symbol &quot;?&quot; reserved for low coverage to be implemented later</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed</span>

<span class="k">def</span><span class="w"> </span><span class="nf">read_sample_sheet</span><span class="p">(</span>

<span class="w">    </span><span class="n">sample_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">genbank_dir</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl_dir</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_sheet</span><span class="o">=</span><span class="s2">&quot;&quot;</span>

<span class="p">):</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Read a sample sheet into a BedmethylItemGroup.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **sample_sheet**</span>

<span class="s2">    &gt; CSV file path (`str`). No header and columns must be in this order: projectname,</span>

<span class="s2">    sample, Genbank name (without extension), bedmethyl file.</span>

<span class="s2">    **genbank_dir**</span>

<span class="s2">    &gt; Directory of the Genbank files (`str`). Default: local directory.</span>

<span class="s2">    **bedmethyl_dir**</span>

<span class="s2">    &gt; Directory of the bedmethyl files (`str`). Default: local directory.</span>

<span class="s2">    **parameter_sheet**</span>

<span class="s2">    &gt; CSV file path (`str`). Use &#39;Parameter&#39;, &#39;Value&#39; header for columns. If a</span>

<span class="s2">    &#39;projectname&#39; is specified, it overwrites the sample sheet value.</span>

<span class="s2">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="c1"># READ PARAMETERS</span>

<span class="w">    </span><span class="n">param_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">parameter_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Value&quot;</span><span class="p">])</span>

<span class="w">    </span><span class="n">parameter_dict</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">dict</span><span class="p">(</span><span class="n">param_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># READ SAMPLES</span>

<span class="w">    </span><span class="n">sample_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># add columnnames</span>

<span class="w">    </span><span class="c1"># CREATE ITEMS</span>

<span class="w">    </span><span class="c1"># We allow Sequeduct to specify the projectname as a command parameter as well;</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="s2">&quot;projectname&quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">:</span>

<span class="w">        </span><span class="c1"># first entry of the first column (contains projectname):</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;projectname&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">]</span>

<span class="w">    </span><span class="c1"># Allow the user to not specify these in the sheet:</span>

<span class="w">    </span><span class="c1"># The default cutoffs are based on preliminary data.</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="s2">&quot;methylated_cutoff&quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">:</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;methylated_cutoff&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.7</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="s2">&quot;unmethylated_cutoff&quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">:</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="p">[</span><span class="s2">&quot;unmethylated_cutoff&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.3</span>

<span class="w">    </span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">index</span><span class="p">,</span><span class="w"> </span><span class="n">row</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">sample_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

<span class="w">        </span><span class="n">genbank_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">        </span><span class="n">genbank_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genbank_dir</span><span class="p">,</span><span class="w"> </span><span class="n">genbank_name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;.gb&quot;</span><span class="p">)</span><span class="w">  </span><span class="c1"># Genbank ext</span>

<span class="w">        </span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;genbank&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genbank_name</span>

<span class="w">        </span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genbank_name</span>

<span class="w">        </span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;DNA&quot;</span>

<span class="w">        </span><span class="n">bed_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">        </span><span class="n">bed_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bedmethyl_dir</span><span class="p">,</span><span class="w"> </span><span class="n">bed_name</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">bed_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="w"> </span><span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_df</span><span class="o">.</span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BEDMETHYL_HEADER</span>

<span class="w">        </span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="p">[</span>

<span class="w">            </span><span class="n">BedmethylItem</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="w"> </span><span class="n">reference</span><span class="o">=</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl</span><span class="o">=</span><span class="n">bed_df</span><span class="p">)</span>

<span class="w">        </span><span class="p">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">    </span><span class="n">bedmethylitemgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BedmethylItemGroup</span><span class="p">(</span>

<span class="w">        </span><span class="n">bedmethylitems</span><span class="o">=</span><span class="n">bedmethylitems</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">=</span><span class="n">parameter_dict</span>

<span class="w">    </span><span class="p">)</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="n">bedmethylitemgroup</span>

<span class="k">def</span><span class="w"> </span><span class="nf">subset_bed_columns</span><span class="p">(</span><span class="n">bed</span><span class="p">):</span>

<span class="w">    </span><span class="n">bed_report</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">[</span><span class="n">columns_for_pdf_report</span><span class="p">]</span>

<span class="w">    </span><span class="c1"># These were designed to be more informative and fit the report:</span>

<span class="w">    </span><span class="n">new_columnnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span>

<span class="w">        </span><span class="s2">&quot;LOC&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;COV&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;% mod&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;MOD&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;STD&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;OTH&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;del&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;fail&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;diff&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;nocall&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;STATUS&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="p">]</span>

<span class="w">    </span><span class="n">bed_report</span><span class="o">.</span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">new_columnnames</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="n">bed_report</span>

<span class="k">class</span><span class="w"> </span><span class="nc">BedmethylItemGroup</span><span class="p">:</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;A group of BedmethylItem instances for reporting.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **bedmethylitems**</span>

<span class="s2">    &gt; A list of BedmethylItem instances.</span>

<span class="s2">    **parameter_dict**</span>

<span class="s2">    &gt; A dictionary of analysis parameters (`dict`).</span>

<span class="s2">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethylitems</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">):</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bedmethylitems</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_dict</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">parameter_dict</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">bedmethylitems</span><span class="p">)</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">perform_all_analysis_in_bedmethylitemgroup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">bedmethylitem</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">bedmethylitems</span><span class="p">:</span>

<span class="w">            </span><span class="n">bedmethylitem</span><span class="o">.</span><span class="n">perform_analysis</span><span class="p">(</span><span class="n">parameter_dict</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_dict</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">comparisons_performed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">True</span>
</code></pre></div>

</details>
<h2 id="variables">Variables</h2>
<div class="codehilite"><pre><span></span><code><span class="n">BEDMETHYL_HEADER</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">COMPLEMENTS</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">DATA_DIR</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">METHYLASES</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">MODIFICATION_CODES</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">columns_for_pdf_report</span>
</code></pre></div>

<h2 id="functions">Functions</h2>
<h3 id="read_sample_sheet">read_sample_sheet</h3>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">read_sample_sheet</span><span class="p">(</span>
    <span class="n">sample_sheet</span><span class="p">,</span>
    <span class="n">genbank_dir</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
    <span class="n">bedmethyl_dir</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
    <span class="n">parameter_sheet</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
<span class="p">)</span>
</code></pre></div>

<p>Read a sample sheet into a BedmethylItemGroup.</p>
<p><strong>Parameters</strong></p>
<p><strong>sample_sheet</strong></p>
<blockquote>
<p>CSV file path (<code>str</code>). No header and columns must be in this order: projectname,
sample, Genbank name (without extension), bedmethyl file.</p>
</blockquote>
<p><strong>genbank_dir</strong></p>
<blockquote>
<p>Directory of the Genbank files (<code>str</code>). Default: local directory.</p>
</blockquote>
<p><strong>bedmethyl_dir</strong></p>
<blockquote>
<p>Directory of the bedmethyl files (<code>str</code>). Default: local directory.</p>
</blockquote>
<p><strong>parameter_sheet</strong></p>
<blockquote>
<p>CSV file path (<code>str</code>). Use 'Parameter', 'Value' header for columns. If a
'projectname' is specified, it overwrites the sample sheet value.</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="n">def</span><span class="w"> </span><span class="n">read_sample_sheet</span><span class="p">(</span>

<span class="w">    </span><span class="n">sample_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">genbank_dir</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl_dir</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_sheet</span><span class="o">=</span><span class="s2">&quot;&quot;</span>

<span class="p">)</span><span class="o">:</span>

<span class="w">    </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Read a sample sheet into a BedmethylItemGroup.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **sample_sheet**</span>

<span class="s2">    &gt; CSV file path (`str`). No header and columns must be in this order: projectname,</span>

<span class="s2">    sample, Genbank name (without extension), bedmethyl file.</span>

<span class="s2">    **genbank_dir**</span>

<span class="s2">    &gt; Directory of the Genbank files (`str`). Default: local directory.</span>

<span class="s2">    **bedmethyl_dir**</span>

<span class="s2">    &gt; Directory of the bedmethyl files (`str`). Default: local directory.</span>

<span class="s2">    **parameter_sheet**</span>

<span class="s2">    &gt; CSV file path (`str`). Use &#39;Parameter&#39;, &#39;Value&#39; header for columns. If a</span>

<span class="s2">    &#39;projectname&#39; is specified, it overwrites the sample sheet value.</span>

<span class="s2">    </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="c1"># READ PARAMETERS</span>

<span class="w">    </span><span class="n">param_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">parameter_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">usecols</span><span class="o">=</span><span class="err">[</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Value&quot;</span><span class="err">]</span><span class="p">)</span>

<span class="w">    </span><span class="n">parameter_dict</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dict</span><span class="p">(</span><span class="n">param_df</span><span class="p">.</span><span class="k">values</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># READ SAMPLES</span>

<span class="w">    </span><span class="n">sample_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_sheet</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="k">None</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># add columnnames</span>

<span class="w">    </span><span class="c1"># CREATE ITEMS</span>

<span class="w">    </span><span class="c1"># We allow Sequeduct to specify the projectname as a command parameter as well;</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="k">not</span><span class="w"> </span><span class="s2">&quot;projectname&quot;</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">:</span>

<span class="w">        </span><span class="c1"># first entry of the first column (contains projectname):</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="err">[</span><span class="s2">&quot;projectname&quot;</span><span class="err">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_df</span><span class="p">.</span><span class="n">iloc</span><span class="err">[</span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="err">]</span>

<span class="w">    </span><span class="c1"># Allow the user to not specify these in the sheet:</span>

<span class="w">    </span><span class="c1"># The default cutoffs are based on preliminary data.</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="k">not</span><span class="w"> </span><span class="s2">&quot;methylated_cutoff&quot;</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">:</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="err">[</span><span class="s2">&quot;methylated_cutoff&quot;</span><span class="err">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.7</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="k">not</span><span class="w"> </span><span class="s2">&quot;unmethylated_cutoff&quot;</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">:</span>

<span class="w">        </span><span class="n">parameter_dict</span><span class="err">[</span><span class="s2">&quot;unmethylated_cutoff&quot;</span><span class="err">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.3</span>

<span class="w">    </span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="k">index</span><span class="p">,</span><span class="w"> </span><span class="k">row</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">sample_df</span><span class="p">.</span><span class="n">iterrows</span><span class="p">()</span><span class="o">:</span>

<span class="w">        </span><span class="n">genbank_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">row</span><span class="err">[</span><span class="mi">2</span><span class="err">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">        </span><span class="n">genbank_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">os</span><span class="p">.</span><span class="k">path</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">genbank_dir</span><span class="p">,</span><span class="w"> </span><span class="n">genbank_name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;.gb&quot;</span><span class="p">)</span><span class="w">  </span><span class="c1"># Genbank ext</span>

<span class="w">        </span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Bio</span><span class="p">.</span><span class="n">SeqIO</span><span class="p">.</span><span class="k">read</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;genbank&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">record</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genbank_name</span>

<span class="w">        </span><span class="n">record</span><span class="p">.</span><span class="k">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genbank_name</span>

<span class="w">        </span><span class="n">record</span><span class="p">.</span><span class="n">annotations</span><span class="err">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="err">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;DNA&quot;</span>

<span class="w">        </span><span class="n">bed_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">row</span><span class="err">[</span><span class="mi">3</span><span class="err">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">        </span><span class="n">bed_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">os</span><span class="p">.</span><span class="k">path</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">bedmethyl_dir</span><span class="p">,</span><span class="w"> </span><span class="n">bed_name</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">bed_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_df</span><span class="p">.</span><span class="k">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BEDMETHYL_HEADER</span>

<span class="w">        </span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="err">[</span>

<span class="w">            </span><span class="n">BedmethylItem</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="k">row</span><span class="err">[</span><span class="mi">1</span><span class="err">]</span><span class="p">,</span><span class="w"> </span><span class="k">reference</span><span class="o">=</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl</span><span class="o">=</span><span class="n">bed_df</span><span class="p">)</span>

<span class="w">        </span><span class="err">]</span><span class="w">  </span><span class="c1"># number specified by the sample sheet format</span>

<span class="w">    </span><span class="n">bedmethylitemgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BedmethylItemGroup</span><span class="p">(</span>

<span class="w">        </span><span class="n">bedmethylitems</span><span class="o">=</span><span class="n">bedmethylitems</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">=</span><span class="n">parameter_dict</span>

<span class="w">    </span><span class="p">)</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="n">bedmethylitemgroup</span>
</code></pre></div>

</details>
<h3 id="subset_bed_columns">subset_bed_columns</h3>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">subset_bed_columns</span><span class="p">(</span>
    <span class="n">bed</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_columns</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span><span class="err">:</span>

<span class="w">    </span><span class="n">bed_report</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">columns_for_pdf_report</span><span class="o">]</span>

<span class="w">    </span><span class="err">#</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">were</span><span class="w"> </span><span class="n">designed</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">informative</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">fit</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="nl">report</span><span class="p">:</span>

<span class="w">    </span><span class="n">new_columnnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>

<span class="n">        &quot;LOC&quot;,</span>

<span class="n">        &quot;Strand&quot;,</span>

<span class="n">        &quot;COV&quot;,</span>

<span class="n">        &quot;% mod&quot;,</span>

<span class="n">        &quot;MOD&quot;,</span>

<span class="n">        &quot;STD&quot;,</span>

<span class="n">        &quot;OTH&quot;,</span>

<span class="n">        &quot;del&quot;,</span>

<span class="n">        &quot;fail&quot;,</span>

<span class="n">        &quot;diff&quot;,</span>

<span class="n">        &quot;nocall&quot;,</span>

<span class="n">        &quot;STATUS&quot;,</span>

<span class="n">    </span><span class="o">]</span>

<span class="w">    </span><span class="n">bed_report</span><span class="p">.</span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">new_columnnames</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="n">bed_report</span>
</code></pre></div>

</details>
<h2 id="classes">Classes</h2>
<h3 id="bedresult">BedResult</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">BedResult</span><span class="p">(</span>
    <span class="n">modification</span><span class="p">,</span>
    <span class="n">bed</span><span class="p">,</span>
    <span class="n">record</span>
<span class="p">)</span>
</code></pre></div>

<p>Results of a bedmethyl table analysis.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nl">BedResult</span><span class="p">:</span>

<span class="w">    </span><span class="ss">&quot;&quot;&quot;Results of a bedmethyl table analysis.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">__init__</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">modification</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">feature_cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">50</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">do</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="k">create</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">there</span><span class="w"> </span><span class="k">are</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">modification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modification</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">mod_abbreviation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span>

<span class="n">            MODIFICATION_CODES[&quot;Code&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">modification</span>

<span class="w">        </span><span class="err">]</span><span class="p">.</span><span class="n">Abbreviation</span><span class="p">.</span><span class="n">iloc</span><span class="o">[</span>

<span class="n">            0</span>

<span class="n">        </span><span class="o">]</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">there</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">entry</span><span class="w"> </span><span class="n">so</span><span class="w"> </span><span class="n">we</span><span class="w"> </span><span class="n">take</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">first</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">mod_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span>

<span class="n">            MODIFICATION_CODES[&quot;Code&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">modification</span>

<span class="w">        </span><span class="err">]</span><span class="p">.</span><span class="n">Name</span><span class="p">.</span><span class="n">iloc</span><span class="o">[</span><span class="n">0</span><span class="o">]</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">mod_chebi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span>

<span class="n">            MODIFICATION_CODES[&quot;Code&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">modification</span>

<span class="w">        </span><span class="err">]</span><span class="p">.</span><span class="n">ChEBI</span><span class="p">.</span><span class="n">iloc</span><span class="o">[</span><span class="n">0</span><span class="o">]</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">unmodified_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MODIFICATION_CODES</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span>

<span class="n">            MODIFICATION_CODES[&quot;Code&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">modification</span>

<span class="w">        </span><span class="err">]</span><span class="p">.</span><span class="n">Unmodified_base</span><span class="p">.</span><span class="n">iloc</span><span class="o">[</span><span class="n">0</span><span class="o">]</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">record</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">FILTER</span><span class="w"> </span><span class="p">(</span><span class="n">ANNOTATED</span><span class="p">)</span><span class="w"> </span><span class="n">RECORD</span><span class="w"> </span><span class="ow">AND</span><span class="w"> </span><span class="k">ADD</span><span class="w"> </span><span class="nl">STATUS</span><span class="p">:</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">labelstart</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span>

<span class="w">            </span><span class="ss">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">unmodified_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">&quot;, strand=&quot;</span>

<span class="w">        </span><span class="p">)</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">unfinished</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">account</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="o">+/-</span><span class="w"> </span><span class="n">strands</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">record</span><span class="p">.</span><span class="nl">features</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;@epijinn&quot;</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">annotated</span><span class="w"> </span><span class="k">by</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">function</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">labelstart</span><span class="p">)</span><span class="err">:</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="k">To</span><span class="w"> </span><span class="n">display</span><span class="w"> </span><span class="n">seq</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">report</span><span class="p">:</span>

<span class="w">                    </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">unmodified_base</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="n">Assign</span><span class="w"> </span><span class="nl">status</span><span class="p">:</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="k">no</span><span class="w"> </span><span class="n">need</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">check</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">strand</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">complement</span><span class="w"> </span><span class="n">won</span><span class="err">&#39;</span><span class="n">t</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">modified</span>

<span class="w">                    </span><span class="n">location_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nc">int</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">location</span><span class="p">.</span><span class="k">start</span><span class="p">)</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="n">expect</span><span class="w"> </span><span class="n">exactly</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="nc">table</span><span class="w"> </span><span class="n">entry</span><span class="w"> </span><span class="n">per</span><span class="w"> </span><span class="n">modified</span><span class="w"> </span><span class="nl">nucleotide</span><span class="p">:</span>

<span class="w">                    </span><span class="n">status_symbol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span>

<span class="n">                        self.bed[&quot;LOC&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">location_start</span>

<span class="w">                    </span><span class="err">]</span><span class="p">.</span><span class="n">STATUS</span><span class="p">.</span><span class="n">iloc</span><span class="o">[</span><span class="n">0</span><span class="o">]</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="k">by</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">custom</span><span class="w"> </span><span class="n">BiopythonTranslator</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="nl">annotation</span><span class="p">:</span>

<span class="w">                    </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;status&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">status_symbol</span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature</span><span class="o">]</span>

<span class="w">        </span><span class="n">record</span><span class="p">.</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_features</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">.</span><span class="n">features</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">feature_cutoff</span><span class="p">:</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">img_created</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">False</span>

<span class="w">        </span><span class="k">else</span><span class="err">:</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">img_created</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">True</span>

<span class="w">            </span><span class="n">fig</span><span class="p">,</span><span class="w"> </span><span class="n">ax1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plt</span><span class="p">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">))</span>

<span class="w">            </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PatternTranslator</span><span class="p">().</span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">            </span><span class="n">graphic_record</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">with_ruler</span><span class="o">=</span><span class="k">False</span><span class="p">,</span><span class="w"> </span><span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fig</span>
</code></pre></div>

</details>
<hr />
<h3 id="bedmethylitem">BedmethylItem</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">BedmethylItem</span><span class="p">(</span>
    <span class="n">sample</span><span class="p">,</span>
    <span class="n">reference</span><span class="p">,</span>
    <span class="n">bedmethyl</span>
<span class="p">)</span>
</code></pre></div>

<p>Analyse a bedmethyl file.</p>
<p><strong>Parameters</strong></p>
<p><strong>sample</strong></p>
<blockquote>
<p>The name of the sample, for example a barcode id (<code>str</code>).</p>
</blockquote>
<p><strong>reference</strong></p>
<blockquote>
<p>A Biopython SeqRecord.</p>
</blockquote>
<p><strong>bedmethyl</strong></p>
<blockquote>
<p>A pandas dataframe of a bedmethyl file for the reference.</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nl">BedmethylItem</span><span class="p">:</span>

<span class="w">    </span><span class="ss">&quot;&quot;&quot;Analyse a bedmethyl file.</span>

<span class="ss">    **Parameters**</span>

<span class="ss">    **sample**</span>

<span class="ss">    &gt; The name of the sample, for example a barcode id (`str`).</span>

<span class="ss">    **reference**</span>

<span class="ss">    &gt; A Biopython SeqRecord.</span>

<span class="ss">    **bedmethyl**</span>

<span class="ss">    &gt; A pandas dataframe of a bedmethyl file for the reference.</span>

<span class="ss">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">__init__</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="p">,</span><span class="w"> </span><span class="n">reference</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethyl</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">sample</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span><span class="p">.</span><span class="n">name</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reference</span><span class="p">.</span><span class="n">id</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bedmethyl</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">reference_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">perform_analysis</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Perform analysis and plot the sequence.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">methylase_str</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;methylases&quot;</span><span class="o">]</span>

<span class="w">        </span><span class="n">methylase</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">METHYLASES</span><span class="o">[</span><span class="n">methylase_str</span><span class="o">]</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">fig</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">plot_record</span><span class="p">()</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">BedResult</span><span class="w"> </span><span class="n">instances</span><span class="p">.</span><span class="w"> </span><span class="k">For</span><span class="w"> </span><span class="n">easy</span><span class="w"> </span><span class="n">reference</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">pug</span><span class="w"> </span><span class="n">template</span><span class="p">.</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">modification</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="p">.</span><span class="n">modified_base_code_and_motif</span><span class="p">.</span><span class="k">unique</span><span class="p">()</span><span class="err">:</span>

<span class="w">            </span><span class="err">#</span><span class="w"> </span><span class="n">RUN</span><span class="w"> </span><span class="n">BED</span><span class="w"> </span><span class="n">SUBSET</span><span class="w"> </span><span class="n">ETC</span>

<span class="w">            </span><span class="n">bed_subtype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="k">mod</span><span class="o">=</span><span class="n">modification</span><span class="p">)</span>

<span class="w">            </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">subset_bed_to_pattern_match</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">bed_subtype</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">bed_binarized</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">binarize_bed</span><span class="p">(</span>

<span class="w">                </span><span class="n">bed_pattern_match</span><span class="p">,</span>

<span class="w">                </span><span class="n">met_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;methylated_cutoff&quot;</span><span class="o">]</span><span class="p">,</span>

<span class="w">                </span><span class="n">nonmet_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;unmethylated_cutoff&quot;</span><span class="o">]</span><span class="p">,</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">final_bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset_bed_columns</span><span class="p">(</span><span class="n">bed_binarized</span><span class="p">)</span>

<span class="w">            </span><span class="n">bedresult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BedResult</span><span class="p">(</span><span class="n">modification</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">final_bed</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="o">=</span><span class="n">annotated_record</span><span class="p">)</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">results</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">bedresult</span><span class="o">]</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">annotate_record</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">record</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">plot_record</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">fig</span><span class="p">,</span><span class="w"> </span><span class="n">ax1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plt</span><span class="p">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">))</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">CustomTranslator</span><span class="p">().</span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">with_ruler</span><span class="o">=</span><span class="k">False</span><span class="p">,</span><span class="w"> </span><span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">fig</span>

<span class="w">    </span><span class="nv">@staticmethod</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="k">mod</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Columnname</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">Bedmethyl</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="n">specification</span>

<span class="w">        </span><span class="n">bed_subtype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">bed[&quot;modified_base_code_and_motif&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="k">mod</span><span class="err">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed_subtype</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_to_base_matches</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">modified_base</span><span class="o">=</span><span class="ss">&quot;C&quot;</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">None</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">optional</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">linking</span><span class="w"> </span><span class="n">multiple</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">subset</span><span class="w"> </span><span class="n">methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modified_base</span><span class="p">.</span><span class="nf">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">mod_base_complement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">COMPLEMENTS</span><span class="o">[</span><span class="n">mod_base</span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">POSITIVE</span><span class="w"> </span><span class="n">STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_positive_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>

<span class="n">            pos</span>

<span class="n">            for pos, base in enumerate(str(self.record.seq.upper()))</span>

<span class="n">            if base == mod_base</span>

<span class="n">        </span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Columnname</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">Bedmethyl</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="nl">specification</span><span class="p">:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_positive_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">NEGATIVE</span><span class="w"> </span><span class="n">STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_negative_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>

<span class="n">            pos</span>

<span class="n">            for pos, base in enumerate(str(self.record.seq.upper()))</span>

<span class="n">            if base == mod_base_complement</span>

<span class="n">        </span><span class="o">]</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_negative_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_basematch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">positive_strand_filter | negative_strand_filter</span><span class="o">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed_basematch</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_to_pattern_match</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">None</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">optional</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">linking</span><span class="w"> </span><span class="n">multiple</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">subset</span><span class="w"> </span><span class="n">methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">methylated_index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="n">index_pos</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="k">sequence</span><span class="o">[</span><span class="n">methylated_index</span><span class="o">]</span><span class="p">.</span><span class="nf">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">record_copy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">copy</span><span class="p">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">annotated_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotate_methylation</span><span class="p">(</span><span class="n">record_copy</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=[</span><span class="n">methylase</span><span class="o">]</span><span class="p">)</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">CREATE</span><span class="w"> </span><span class="n">LIST</span><span class="w"> </span><span class="k">OF</span><span class="w"> </span><span class="n">POSITIONS</span>

<span class="w">        </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">label_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">&quot;, strand=1)&quot;</span>

<span class="w">        </span><span class="n">label_neg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">&quot;, strand=-1)&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">annotated_record</span><span class="p">.</span><span class="nl">features</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;@epijinn&quot;</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">annotated</span><span class="w"> </span><span class="k">by</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">above</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nl">label_pos</span><span class="p">:</span>

<span class="w">                    </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature.location.start</span><span class="o">]</span>

<span class="w">                </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nl">label_neg</span><span class="p">:</span>

<span class="w">                    </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature.location.start</span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">SUBSET</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="n">POSITIONS</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Columnname</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">Bedmethyl</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="nl">specification</span><span class="p">:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">positive_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">negative_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">positive_strand_filter | negative_strand_filter</span><span class="o">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span>

<span class="w">    </span><span class="nv">@staticmethod</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">binarize_bed</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="n">met_cutoff</span><span class="p">,</span><span class="w"> </span><span class="n">nonmet_cutoff</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;status&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;U&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">prefill</span><span class="w"> </span><span class="n">undetermined</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="k">index</span><span class="p">,</span><span class="w"> </span><span class="k">row</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">iterrows</span><span class="p">()</span><span class="err">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="k">row</span><span class="o">[</span><span class="n">&quot;percent_modified&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="nc">float</span><span class="p">(</span><span class="n">met_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="err">:</span>

<span class="w">                </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">index, &quot;status&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;1&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">methylated</span><span class="p">,</span><span class="w"> </span><span class="n">symbol</span><span class="w"> </span><span class="n">may</span><span class="w"> </span><span class="n">change</span>

<span class="w">            </span><span class="n">elif</span><span class="w"> </span><span class="k">row</span><span class="o">[</span><span class="n">&quot;percent_modified&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nc">float</span><span class="p">(</span><span class="n">nonmet_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="err">:</span>

<span class="w">                </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">index, &quot;status&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;0&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">unmethylated</span>

<span class="w">            </span><span class="err">#</span><span class="w"> </span><span class="n">symbol</span><span class="w"> </span><span class="ss">&quot;?&quot;</span><span class="w"> </span><span class="n">reserved</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">low</span><span class="w"> </span><span class="n">coverage</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">implemented</span><span class="w"> </span><span class="n">later</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed</span>
</code></pre></div>

</details>
<hr />
<h4 id="static-methods">Static methods</h4>
<h4 id="binarize_bed">binarize_bed</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">binarize_bed</span><span class="p">(</span>
    <span class="n">bed</span><span class="p">,</span>
    <span class="n">met_cutoff</span><span class="p">,</span>
    <span class="n">nonmet_cutoff</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="err">@</span><span class="nx">staticmethod</span>

<span class="w">    </span><span class="nx">def</span><span class="w"> </span><span class="nx">binarize_bed</span><span class="p">(</span><span class="nx">bed</span><span class="p">,</span><span class="w"> </span><span class="nx">met_cutoff</span><span class="p">,</span><span class="w"> </span><span class="nx">nonmet_cutoff</span><span class="p">):</span>

<span class="w">        </span><span class="nx">bed</span><span class="p">[</span><span class="s">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;U&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="nx">prefill</span><span class="w"> </span><span class="nx">undetermined</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="nx">index</span><span class="p">,</span><span class="w"> </span><span class="nx">row</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">bed</span><span class="p">.</span><span class="nx">iterrows</span><span class="p">():</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nx">row</span><span class="p">[</span><span class="s">&quot;percent_modified&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="nx">float</span><span class="p">(</span><span class="nx">met_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">:</span>

<span class="w">                </span><span class="nx">bed</span><span class="p">.</span><span class="nx">loc</span><span class="p">[</span><span class="nx">index</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;1&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="nx">methylated</span><span class="p">,</span><span class="w"> </span><span class="nx">symbol</span><span class="w"> </span><span class="nx">may</span><span class="w"> </span><span class="nx">change</span>

<span class="w">            </span><span class="nx">elif</span><span class="w"> </span><span class="nx">row</span><span class="p">[</span><span class="s">&quot;percent_modified&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nx">float</span><span class="p">(</span><span class="nx">nonmet_cutoff</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">:</span>

<span class="w">                </span><span class="nx">bed</span><span class="p">.</span><span class="nx">loc</span><span class="p">[</span><span class="nx">index</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;status&quot;</span><span class="p">]</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;0&quot;</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="nx">unmethylated</span>

<span class="w">            </span><span class="err">#</span><span class="w"> </span><span class="nx">symbol</span><span class="w"> </span><span class="s">&quot;?&quot;</span><span class="w"> </span><span class="nx">reserved</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">low</span><span class="w"> </span><span class="nx">coverage</span><span class="w"> </span><span class="nx">to</span><span class="w"> </span><span class="nx">be</span><span class="w"> </span><span class="nx">implemented</span><span class="w"> </span><span class="nx">later</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="nx">bed</span>
</code></pre></div>

</details>
<h4 id="subset_bed_to_mod_subtype">subset_bed_to_mod_subtype</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">subset_bed_to_mod_subtype</span><span class="p">(</span>
    <span class="n">bed</span><span class="p">,</span>
    <span class="n">mod</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="err">@</span><span class="nx">staticmethod</span>

<span class="w">    </span><span class="nx">def</span><span class="w"> </span><span class="nx">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="nx">bed</span><span class="p">,</span><span class="w"> </span><span class="nx">mod</span><span class="p">):</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="nx">Columnname</span><span class="w"> </span><span class="nx">from</span><span class="w"> </span><span class="nx">Bedmethyl</span><span class="w"> </span><span class="nx">file</span><span class="w"> </span><span class="nx">specification</span>

<span class="w">        </span><span class="nx">bed_subtype</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">bed</span><span class="p">.</span><span class="nx">loc</span><span class="p">[</span><span class="nx">bed</span><span class="p">[</span><span class="s">&quot;modified_base_code_and_motif&quot;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nx">mod</span><span class="p">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="nx">bed_subtype</span>
</code></pre></div>

</details>
<h4 id="methods">Methods</h4>
<h4 id="annotate_record">annotate_record</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_record</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def annotate_record(self):

        return self.record
</code></pre></div>

</details>
<h4 id="perform_analysis">perform_analysis</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">perform_analysis</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">parameter_dict</span>
<span class="p">)</span>
</code></pre></div>

<p>Perform analysis and plot the sequence.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">perform_analysis</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Perform analysis and plot the sequence.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">methylase_str</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;methylases&quot;</span><span class="o">]</span>

<span class="w">        </span><span class="n">methylase</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">METHYLASES</span><span class="o">[</span><span class="n">methylase_str</span><span class="o">]</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">fig</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">plot_record</span><span class="p">()</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">BedResult</span><span class="w"> </span><span class="n">instances</span><span class="p">.</span><span class="w"> </span><span class="k">For</span><span class="w"> </span><span class="n">easy</span><span class="w"> </span><span class="n">reference</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">pug</span><span class="w"> </span><span class="n">template</span><span class="p">.</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">modification</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="p">.</span><span class="n">modified_base_code_and_motif</span><span class="p">.</span><span class="k">unique</span><span class="p">()</span><span class="err">:</span>

<span class="w">            </span><span class="err">#</span><span class="w"> </span><span class="n">RUN</span><span class="w"> </span><span class="n">BED</span><span class="w"> </span><span class="n">SUBSET</span><span class="w"> </span><span class="n">ETC</span>

<span class="w">            </span><span class="n">bed_subtype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">subset_bed_to_mod_subtype</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">bed</span><span class="p">,</span><span class="w"> </span><span class="k">mod</span><span class="o">=</span><span class="n">modification</span><span class="p">)</span>

<span class="w">            </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">subset_bed_to_pattern_match</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">bed_subtype</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">bed_binarized</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">binarize_bed</span><span class="p">(</span>

<span class="w">                </span><span class="n">bed_pattern_match</span><span class="p">,</span>

<span class="w">                </span><span class="n">met_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;methylated_cutoff&quot;</span><span class="o">]</span><span class="p">,</span>

<span class="w">                </span><span class="n">nonmet_cutoff</span><span class="o">=</span><span class="n">parameter_dict</span><span class="o">[</span><span class="n">&quot;unmethylated_cutoff&quot;</span><span class="o">]</span><span class="p">,</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">final_bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset_bed_columns</span><span class="p">(</span><span class="n">bed_binarized</span><span class="p">)</span>

<span class="w">            </span><span class="n">bedresult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BedResult</span><span class="p">(</span><span class="n">modification</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="n">final_bed</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="o">=</span><span class="n">annotated_record</span><span class="p">)</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">results</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">bedresult</span><span class="o">]</span>
</code></pre></div>

</details>
<h4 id="plot_record">plot_record</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">plot_record</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def plot_record(self):

        fig, ax1 = plt.subplots(1, 1, figsize=(8, 2))

        graphic_record = CustomTranslator().translate_record(self.record)

        graphic_record.plot(ax=ax1, with_ruler=False, strand_in_label_threshold=4)

        return fig
</code></pre></div>

</details>
<h4 id="subset_bed_to_base_matches">subset_bed_to_base_matches</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">subset_bed_to_base_matches</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">bed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">modified_base</span><span class="o">=</span><span class="s1">&#39;C&#39;</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_to_base_matches</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">modified_base</span><span class="o">=</span><span class="ss">&quot;C&quot;</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">None</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">optional</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">linking</span><span class="w"> </span><span class="n">multiple</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">subset</span><span class="w"> </span><span class="n">methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modified_base</span><span class="p">.</span><span class="nf">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">mod_base_complement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">COMPLEMENTS</span><span class="o">[</span><span class="n">mod_base</span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">POSITIVE</span><span class="w"> </span><span class="n">STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_positive_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>

<span class="n">            pos</span>

<span class="n">            for pos, base in enumerate(str(self.record.seq.upper()))</span>

<span class="n">            if base == mod_base</span>

<span class="n">        </span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Columnname</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">Bedmethyl</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="nl">specification</span><span class="p">:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_positive_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">NEGATIVE</span><span class="w"> </span><span class="n">STRAND</span>

<span class="w">        </span><span class="n">matching_positions_on_negative_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>

<span class="n">            pos</span>

<span class="n">            for pos, base in enumerate(str(self.record.seq.upper()))</span>

<span class="n">            if base == mod_base_complement</span>

<span class="n">        </span><span class="o">]</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">matching_positions_on_negative_strand</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_basematch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">positive_strand_filter | negative_strand_filter</span><span class="o">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">bed_basematch</span>
</code></pre></div>

</details>
<h4 id="subset_bed_to_pattern_match">subset_bed_to_pattern_match</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">subset_bed_to_pattern_match</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">methylase</span><span class="p">,</span>
    <span class="n">bed</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">subset_bed_to_pattern_match</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">,</span><span class="w"> </span><span class="n">bed</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">None</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">optional</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">linking</span><span class="w"> </span><span class="n">multiple</span><span class="w"> </span><span class="n">bed</span><span class="w"> </span><span class="n">subset</span><span class="w"> </span><span class="n">methods</span>

<span class="w">            </span><span class="n">bed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bed</span>

<span class="w">        </span><span class="n">methylated_index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="n">index_pos</span>

<span class="w">        </span><span class="n">mod_base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="k">sequence</span><span class="o">[</span><span class="n">methylated_index</span><span class="o">]</span><span class="p">.</span><span class="nf">upper</span><span class="p">()</span>

<span class="w">        </span><span class="n">record_copy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">copy</span><span class="p">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">annotated_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotate_methylation</span><span class="p">(</span><span class="n">record_copy</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=[</span><span class="n">methylase</span><span class="o">]</span><span class="p">)</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">CREATE</span><span class="w"> </span><span class="n">LIST</span><span class="w"> </span><span class="k">OF</span><span class="w"> </span><span class="n">POSITIONS</span>

<span class="w">        </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">label_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">&quot;, strand=1)&quot;</span>

<span class="w">        </span><span class="n">label_neg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">mod_base</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">&quot;, strand=-1)&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">annotated_record</span><span class="p">.</span><span class="nl">features</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;@epijinn&quot;</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">annotated</span><span class="w"> </span><span class="k">by</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">above</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nl">label_pos</span><span class="p">:</span>

<span class="w">                    </span><span class="n">positive_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature.location.start</span><span class="o">]</span>

<span class="w">                </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">&quot;label&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nl">label_neg</span><span class="p">:</span>

<span class="w">                    </span><span class="n">negative_strand_locations</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature.location.start</span><span class="o">]</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">SUBSET</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="n">POSITIONS</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Columnname</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">Bedmethyl</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="nl">specification</span><span class="p">:</span>

<span class="w">        </span><span class="n">positive_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">positive_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;+&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">negative_strand_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;start_position&quot;</span><span class="o">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span>

<span class="w">            </span><span class="n">negative_strand_locations</span>

<span class="w">        </span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed</span><span class="o">[</span><span class="n">&quot;strand&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;-&quot;</span><span class="p">)</span>

<span class="w">        </span><span class="n">bed_pattern_match</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed</span><span class="p">.</span><span class="n">loc</span><span class="o">[</span><span class="n">positive_strand_filter | negative_strand_filter</span><span class="o">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">annotated_record</span><span class="p">,</span><span class="w"> </span><span class="n">bed_pattern_match</span>
</code></pre></div>

</details>
<h3 id="bedmethylitemgroup">BedmethylItemGroup</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">BedmethylItemGroup</span><span class="p">(</span>
    <span class="n">bedmethylitems</span><span class="p">,</span>
    <span class="n">parameter_dict</span>
<span class="p">)</span>
</code></pre></div>

<p>A group of BedmethylItem instances for reporting.</p>
<p><strong>Parameters</strong></p>
<p><strong>bedmethylitems</strong></p>
<blockquote>
<p>A list of BedmethylItem instances.</p>
</blockquote>
<p><strong>parameter_dict</strong></p>
<blockquote>
<p>A dictionary of analysis parameters (<code>dict</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="n">class</span><span class="w"> </span><span class="n">BedmethylItemGroup</span><span class="o">:</span>

<span class="w">    </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">A group of BedmethylItem instances for reporting.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **bedmethylitems**</span>

<span class="s2">    &gt; A list of BedmethylItem instances.</span>

<span class="s2">    **parameter_dict**</span>

<span class="s2">    &gt; A dictionary of analysis parameters (`dict`).</span>

<span class="s2">    </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">__init__</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">bedmethylitems</span><span class="p">,</span><span class="w"> </span><span class="n">parameter_dict</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">bedmethylitems</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bedmethylitems</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">parameter_dict</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">parameter_dict</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">number_of_samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">bedmethylitems</span><span class="p">)</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">perform_all_analysis_in_bedmethylitemgroup</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">bedmethylitem</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">bedmethylitems</span><span class="o">:</span>

<span class="w">            </span><span class="n">bedmethylitem</span><span class="p">.</span><span class="n">perform_analysis</span><span class="p">(</span><span class="n">parameter_dict</span><span class="o">=</span><span class="n">self</span><span class="p">.</span><span class="n">parameter_dict</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">comparisons_performed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="no">True</span>
</code></pre></div>

</details>
<hr />
<h4 id="methods_1">Methods</h4>
<h4 id="perform_all_analysis_in_bedmethylitemgroup">perform_all_analysis_in_bedmethylitemgroup</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">perform_all_analysis_in_bedmethylitemgroup</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def perform_all_analysis_in_bedmethylitemgroup(self):

        for bedmethylitem in self.bedmethylitems:

            bedmethylitem.perform_analysis(parameter_dict=self.parameter_dict)

        self.comparisons_performed = True
</code></pre></div>

</details>
<h3 id="customtranslator">CustomTranslator</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">CustomTranslator</span><span class="p">(</span>
    <span class="n">features_filters</span><span class="o">=</span><span class="p">(),</span>
    <span class="n">features_properties</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Custom translator used for the main plot.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="n">CustomTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="p">.</span><span class="n">BiopythonTranslator</span><span class="p">)</span><span class="err">:</span>

<span class="w">    </span><span class="ss">&quot;&quot;&quot;Custom translator used for the main plot.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Display only selected features.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">8</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">good</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">display</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Keep</span><span class="w"> </span><span class="n">longest</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="nl">n</span><span class="p">:</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span>

<span class="w">                </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">len(feature.location)</span><span class="o">]</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="p">.</span><span class="n">sort</span><span class="p">(</span><span class="nf">reverse</span><span class="o">=</span><span class="k">True</span><span class="p">)</span>

<span class="w">            </span><span class="n">max_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature_lengths</span><span class="o">[</span><span class="n">n</span><span class="o">]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">location</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="nl">max_length</span><span class="p">:</span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature</span><span class="o">]</span>

<span class="w">        </span><span class="k">else</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="k">no</span><span class="w"> </span><span class="n">need</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">do</span><span class="w"> </span><span class="n">anything</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">enough</span><span class="w"> </span><span class="n">features</span>

<span class="w">            </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">filtered_features</span>
</code></pre></div>

</details>
<hr />
<h4 id="ancestors-in-mro">Ancestors (in MRO)</h4>
<ul>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslator.BiopythonTranslator</li>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslatorBase.BiopythonTranslatorBase</li>
</ul>
<h4 id="class-variables">Class variables</h4>
<div class="codehilite"><pre><span></span><code><span class="n">default_feature_color</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">graphic_record_parameters</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">ignored_features_types</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">label_fields</span>
</code></pre></div>

<h4 id="static-methods_1">Static methods</h4>
<h4 id="quick_class_plot">quick_class_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_class_plot</span><span class="p">(</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="nd">@classmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">quick_class_plot</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="s2">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="s2">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="s2">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="s2">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="s2">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">cls</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">ax</span><span class="p">,</span><span class="w"> </span><span class="n">_</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="methods_2">Methods</h4>
<h4 id="compute_feature_box_color">compute_feature_box_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_color for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_box_color(self, feature):

        &quot;&quot;&quot;Compute a box_color for this feature.&quot;&quot;&quot;

        return &quot;auto&quot;
</code></pre></div>

</details>
<h4 id="compute_feature_box_linewidth">compute_feature_box_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_linewidth for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_box_linewidth(self, feature):

        &quot;&quot;&quot;Compute a box_linewidth for this feature.&quot;&quot;&quot;

        return 0.3
</code></pre></div>

</details>
<h4 id="compute_feature_color">compute_feature_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a color for this feature.</p>
<p>If the feature has a <code>color</code> qualifier it will be used. Otherwise,
the classe's <code>default_feature_color</code> is used.</p>
<p>To change the behaviour, create a subclass of <code>BiopythonTranslator</code>
and overwrite this method.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_color</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Compute a color for this feature.</span>

<span class="s2">        If the feature has a ``color`` qualifier it will be used. Otherwise,</span>

<span class="s2">        the classe&#39;s ``default_feature_color`` is used.</span>

<span class="s2">        To change the behaviour, create a subclass of ``BiopythonTranslator``</span>

<span class="s2">        and overwrite this method.</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="s2">&quot;color&quot;</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">:</span>

<span class="w">            </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="err">[</span><span class="s2">&quot;color&quot;</span><span class="err">]</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">color</span><span class="err">[</span><span class="mi">0</span><span class="err">]</span><span class="p">,</span><span class="w"> </span><span class="n">str</span><span class="p">)</span><span class="o">:</span>

<span class="w">                </span><span class="k">return</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="err">[</span><span class="s2">&quot;color&quot;</span><span class="err">]</span><span class="p">)</span>

<span class="w">            </span><span class="k">else</span><span class="o">:</span>

<span class="w">                </span><span class="k">return</span><span class="w"> </span><span class="n">color</span>

<span class="w">        </span><span class="k">else</span><span class="o">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">default_feature_color</span>
</code></pre></div>

</details>
<h4 id="compute_feature_fontdict">compute_feature_fontdict</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_fontdict</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a font dict for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_fontdict(self, feature):

        &quot;&quot;&quot;Compute a font dict for this feature.&quot;&quot;&quot;

        return None
</code></pre></div>

</details>
<h4 id="compute_feature_html">compute_feature_html</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_html</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Gets the 'label' of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_html(self, feature):

        &quot;&quot;&quot;Gets the &#39;label&#39; of the feature.&quot;&quot;&quot;

        return self.compute_feature_label(feature)
</code></pre></div>

</details>
<h4 id="compute_feature_label">compute_feature_label</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the label of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_label</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Compute the label of the feature.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">type</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">label_fields</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="err">:</span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span>

<span class="w">                </span><span class="k">break</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">list</span><span class="p">)</span><span class="err">:</span>

<span class="w">            </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;|&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">label</span>
</code></pre></div>

</details>
<h4 id="compute_feature_label_link_color">compute_feature_label_link_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label_link_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the color of the line linking the label to its feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_label_link_color(self, feature):

        &quot;&quot;&quot;Compute the color of the line linking the label to its feature.&quot;&quot;&quot;

        return &quot;black&quot;
</code></pre></div>

</details>
<h4 id="compute_feature_legend_text">compute_feature_legend_text</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_legend_text</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_legend_text(self, feature):

        return None
</code></pre></div>

</details>
<h4 id="compute_feature_linewidth">compute_feature_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the edge width of the feature's arrow/rectangle.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_linewidth(self, feature):

        &quot;&quot;&quot;Compute the edge width of the feature&#39;s arrow/rectangle.&quot;&quot;&quot;

        return 1.0
</code></pre></div>

</details>
<h4 id="compute_filtered_features">compute_filtered_features</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_filtered_features</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">features</span>
<span class="p">)</span>
</code></pre></div>

<p>Display only selected features.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Display only selected features.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">8</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">good</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">display</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Keep</span><span class="w"> </span><span class="n">longest</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="nl">n</span><span class="p">:</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span>

<span class="w">                </span><span class="n">feature_lengths</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">len(feature.location)</span><span class="o">]</span>

<span class="w">            </span><span class="n">feature_lengths</span><span class="p">.</span><span class="n">sort</span><span class="p">(</span><span class="nf">reverse</span><span class="o">=</span><span class="k">True</span><span class="p">)</span>

<span class="w">            </span><span class="n">max_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature_lengths</span><span class="o">[</span><span class="n">n</span><span class="o">]</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">location</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="nl">max_length</span><span class="p">:</span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature</span><span class="o">]</span>

<span class="w">        </span><span class="k">else</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="k">no</span><span class="w"> </span><span class="n">need</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">do</span><span class="w"> </span><span class="n">anything</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">enough</span><span class="w"> </span><span class="n">features</span>

<span class="w">            </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">features</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">filtered_features</span>
</code></pre></div>

</details>
<h4 id="quick_plot">quick_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_plot</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">quick_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="s2">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="s2">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="s2">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="s2">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="s2">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">ax</span><span class="p">,</span><span class="w"> </span><span class="n">_</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="translate_feature">translate_feature</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_feature</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Translate a Biopython feature into a Dna Features Viewer feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def translate_feature(self, feature):

        &quot;&quot;&quot;Translate a Biopython feature into a Dna Features Viewer feature.&quot;&quot;&quot;

        properties = dict(

            label=self.compute_feature_label(feature),

            color=self.compute_feature_color(feature),

            html=self.compute_feature_html(feature),

            fontdict=self.compute_feature_fontdict(feature),

            box_linewidth=self.compute_feature_box_linewidth(feature),

            box_color=self.compute_feature_box_color(feature),

            linewidth=self.compute_feature_linewidth(feature),

            label_link_color=self.compute_feature_label_link_color(feature),

            legend_text=self.compute_feature_legend_text(feature),

        )

        if self.features_properties is not None:

            other_properties = self.features_properties

            if hasattr(other_properties, &quot;__call__&quot;):

                other_properties = other_properties(feature)

            properties.update(other_properties)

        return GraphicFeature(

            start=feature.location.start,

            end=feature.location.end,

            strand=feature.location.strand,

            **properties

        )
</code></pre></div>

</details>
<h4 id="translate_record">translate_record</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_record</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">record_class</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a new GraphicRecord from a BioPython Record object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>record</td>
<td>None</td>
<td>A BioPython Record object or the path to a Genbank or a GFF file.</td>
<td>None</td>
</tr>
<tr>
<td>record_class</td>
<td>None</td>
<td>The graphic record class to use, e.g. GraphicRecord (default) or<br>CircularGraphicRecord. Strings 'circular' and 'linear' can also be<br>provided.</td>
<td>None</td>
</tr>
<tr>
<td>filetype</td>
<td>None</td>
<td>Used only when a Genbank or a GFF file is provided; one of "genbank"<br>or "gff" to be used. Default None infers from file extension.</td>
<td>None</td>
</tr>
</tbody>
</table>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">record_class</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Create a new GraphicRecord from a BioPython Record object.</span>

<span class="ss">        Parameters</span>

<span class="ss">        ----------</span>

<span class="ss">        record</span>

<span class="ss">          A BioPython Record object or the path to a Genbank or a GFF file.</span>

<span class="ss">        record_class</span>

<span class="ss">          The graphic record class to use, e.g. GraphicRecord (default) or</span>

<span class="ss">          CircularGraphicRecord. Strings &#39;circular&#39; and &#39;linear&#39; can also be</span>

<span class="ss">          provided.</span>

<span class="ss">        filetype</span>

<span class="ss">          Used only when a Genbank or a GFF file is provided; one of &quot;</span><span class="n">genbank</span><span class="ss">&quot;</span>

<span class="ss">          or &quot;</span><span class="n">gff</span><span class="ss">&quot; to be used. Default None infers from file extension.</span>

<span class="ss">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">classes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">{</span>

<span class="w">            </span><span class="ss">&quot;linear&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span>

<span class="w">            </span><span class="ss">&quot;circular&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">CircularGraphicRecord</span><span class="p">,</span>

<span class="w">            </span><span class="k">None</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span>

<span class="w">        </span><span class="err">}</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">record_class</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">classes</span><span class="p">:</span>

<span class="w">            </span><span class="n">record_class</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">classes</span><span class="o">[</span><span class="n">record_class</span><span class="o">]</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="nf">str</span><span class="p">)</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="n">hasattr</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;read&quot;</span><span class="p">)</span><span class="err">:</span>

<span class="w">            </span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">load_record</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="n">filetype</span><span class="p">)</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">features</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">record_class</span><span class="p">(</span>

<span class="w">            </span><span class="n">sequence_length</span><span class="o">=</span><span class="nf">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span>

<span class="w">            </span><span class="k">sequence</span><span class="o">=</span><span class="nf">str</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">.</span><span class="n">defined</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">None</span><span class="p">,</span>

<span class="w">            </span><span class="n">features</span><span class="o">=[</span>

<span class="n">                self.translate_feature(feature)</span>

<span class="n">                for feature in filtered_features</span>

<span class="n">                if feature.location is not None</span>

<span class="n">            </span><span class="o">]</span><span class="p">,</span>

<span class="w">            </span><span class="o">**</span><span class="n">self</span><span class="p">.</span><span class="n">graphic_record_parameters</span>

<span class="w">        </span><span class="p">)</span>
</code></pre></div>

</details>
<h3 id="patterntranslator">PatternTranslator</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">PatternTranslator</span><span class="p">(</span>
    <span class="n">features_filters</span><span class="o">=</span><span class="p">(),</span>
    <span class="n">features_properties</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Custom translator used for the pattern annotation plot.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="n">PatternTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="p">.</span><span class="n">BiopythonTranslator</span><span class="p">)</span><span class="err">:</span>

<span class="w">    </span><span class="ss">&quot;&quot;&quot;Custom translator used for the pattern annotation plot.&quot;&quot;&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_color</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">status</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;status&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">Values</span><span class="w"> </span><span class="k">are</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">binarize_bed</span><span class="p">()</span><span class="w"> </span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;1&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;red&quot;</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;U&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;yellow&quot;</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;0&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;grey&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">implemented</span><span class="w"> </span><span class="nl">later</span><span class="p">:</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;?&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;tab:cyan&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">There</span><span class="w"> </span><span class="n">should</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="k">no</span><span class="w"> </span><span class="n">other</span><span class="w"> </span><span class="n">options</span><span class="p">.</span>
</code></pre></div>

</details>
<hr />
<h4 id="ancestors-in-mro_1">Ancestors (in MRO)</h4>
<ul>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslator.BiopythonTranslator</li>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslatorBase.BiopythonTranslatorBase</li>
</ul>
<h4 id="class-variables_1">Class variables</h4>
<div class="codehilite"><pre><span></span><code><span class="n">default_feature_color</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">graphic_record_parameters</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">ignored_features_types</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">label_fields</span>
</code></pre></div>

<h4 id="static-methods_2">Static methods</h4>
<h4 id="quick_class_plot_1">quick_class_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_class_plot</span><span class="p">(</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="nd">@classmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">quick_class_plot</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="s2">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="s2">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="s2">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="s2">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="s2">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">cls</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">ax</span><span class="p">,</span><span class="w"> </span><span class="n">_</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="methods_3">Methods</h4>
<h4 id="compute_feature_box_color_1">compute_feature_box_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_color for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_box_color(self, feature):

        &quot;&quot;&quot;Compute a box_color for this feature.&quot;&quot;&quot;

        return &quot;auto&quot;
</code></pre></div>

</details>
<h4 id="compute_feature_box_linewidth_1">compute_feature_box_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_linewidth for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_box_linewidth(self, feature):

        &quot;&quot;&quot;Compute a box_linewidth for this feature.&quot;&quot;&quot;

        return 0.3
</code></pre></div>

</details>
<h4 id="compute_feature_color_1">compute_feature_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a color for this feature.</p>
<p>If the feature has a <code>color</code> qualifier it will be used. Otherwise,
the classe's <code>default_feature_color</code> is used.</p>
<p>To change the behaviour, create a subclass of <code>BiopythonTranslator</code>
and overwrite this method.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_color</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">status</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;status&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">Values</span><span class="w"> </span><span class="k">are</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">binarize_bed</span><span class="p">()</span><span class="w"> </span><span class="err">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;1&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;red&quot;</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;U&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;yellow&quot;</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;0&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;grey&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">implemented</span><span class="w"> </span><span class="nl">later</span><span class="p">:</span>

<span class="w">        </span><span class="n">elif</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">status</span><span class="o">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="ss">&quot;?&quot;</span><span class="err">:</span>

<span class="w">            </span><span class="k">return</span><span class="w"> </span><span class="ss">&quot;tab:cyan&quot;</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">There</span><span class="w"> </span><span class="n">should</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="k">no</span><span class="w"> </span><span class="n">other</span><span class="w"> </span><span class="n">options</span><span class="p">.</span>
</code></pre></div>

</details>
<h4 id="compute_feature_fontdict_1">compute_feature_fontdict</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_fontdict</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a font dict for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_fontdict(self, feature):

        &quot;&quot;&quot;Compute a font dict for this feature.&quot;&quot;&quot;

        return None
</code></pre></div>

</details>
<h4 id="compute_feature_html_1">compute_feature_html</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_html</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Gets the 'label' of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_html(self, feature):

        &quot;&quot;&quot;Gets the &#39;label&#39; of the feature.&quot;&quot;&quot;

        return self.compute_feature_label(feature)
</code></pre></div>

</details>
<h4 id="compute_feature_label_1">compute_feature_label</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the label of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_label</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Compute the label of the feature.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">type</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">label_fields</span><span class="p">:</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="err">:</span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span>

<span class="w">                </span><span class="k">break</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">list</span><span class="p">)</span><span class="err">:</span>

<span class="w">            </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;|&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">label</span>
</code></pre></div>

</details>
<h4 id="compute_feature_label_link_color_1">compute_feature_label_link_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label_link_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the color of the line linking the label to its feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_label_link_color(self, feature):

        &quot;&quot;&quot;Compute the color of the line linking the label to its feature.&quot;&quot;&quot;

        return &quot;black&quot;
</code></pre></div>

</details>
<h4 id="compute_feature_legend_text_1">compute_feature_legend_text</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_legend_text</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_legend_text(self, feature):

        return None
</code></pre></div>

</details>
<h4 id="compute_feature_linewidth_1">compute_feature_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the edge width of the feature's arrow/rectangle.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def compute_feature_linewidth(self, feature):

        &quot;&quot;&quot;Compute the edge width of the feature&#39;s arrow/rectangle.&quot;&quot;&quot;

        return 1.0
</code></pre></div>

</details>
<h4 id="compute_filtered_features_1">compute_filtered_features</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_filtered_features</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">features</span>
<span class="p">)</span>
</code></pre></div>

<p>Return the list of features minus the ignored ones.</p>
<p>By the method keeps any feature whose type is not in
ignored_features_types and for which all filter(f) pass.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="nx">def</span><span class="w"> </span><span class="nx">compute_filtered_features</span><span class="p">(</span><span class="kp">self</span><span class="p">,</span><span class="w"> </span><span class="nx">features</span><span class="p">):</span>

<span class="w">        </span><span class="s">&quot;&quot;&quot;Return the list of features minus the ignored ones.</span>

<span class="s">        By the method keeps any feature whose type is not in</span>

<span class="s">        ignored_features_types and for which all filter(f) pass.</span>

<span class="s">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="p">[</span>

<span class="w">            </span><span class="nx">f</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="nx">f</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">features</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nx">all</span><span class="p">([</span><span class="nx">fl</span><span class="p">(</span><span class="nx">f</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">fl</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="kp">self</span><span class="p">.</span><span class="nx">features_filters</span><span class="p">])</span>

<span class="w">            </span><span class="k">and</span><span class="w"> </span><span class="nx">f</span><span class="p">.</span><span class="k">type</span><span class="w"> </span><span class="k">not</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="kp">self</span><span class="p">.</span><span class="nx">ignored_features_types</span>

<span class="w">        </span><span class="p">]</span>
</code></pre></div>

</details>
<h4 id="quick_plot_1">quick_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_plot</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">quick_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="s2">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="s2">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="s2">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="s2">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="s2">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="w">        </span><span class="n">ax</span><span class="p">,</span><span class="w"> </span><span class="n">_</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span><span class="w"> </span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="translate_feature_1">translate_feature</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_feature</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Translate a Biopython feature into a Dna Features Viewer feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def translate_feature(self, feature):

        &quot;&quot;&quot;Translate a Biopython feature into a Dna Features Viewer feature.&quot;&quot;&quot;

        properties = dict(

            label=self.compute_feature_label(feature),

            color=self.compute_feature_color(feature),

            html=self.compute_feature_html(feature),

            fontdict=self.compute_feature_fontdict(feature),

            box_linewidth=self.compute_feature_box_linewidth(feature),

            box_color=self.compute_feature_box_color(feature),

            linewidth=self.compute_feature_linewidth(feature),

            label_link_color=self.compute_feature_label_link_color(feature),

            legend_text=self.compute_feature_legend_text(feature),

        )

        if self.features_properties is not None:

            other_properties = self.features_properties

            if hasattr(other_properties, &quot;__call__&quot;):

                other_properties = other_properties(feature)

            properties.update(other_properties)

        return GraphicFeature(

            start=feature.location.start,

            end=feature.location.end,

            strand=feature.location.strand,

            **properties

        )
</code></pre></div>

</details>
<h4 id="translate_record_1">translate_record</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_record</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">record_class</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a new GraphicRecord from a BioPython Record object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>record</td>
<td>None</td>
<td>A BioPython Record object or the path to a Genbank or a GFF file.</td>
<td>None</td>
</tr>
<tr>
<td>record_class</td>
<td>None</td>
<td>The graphic record class to use, e.g. GraphicRecord (default) or<br>CircularGraphicRecord. Strings 'circular' and 'linear' can also be<br>provided.</td>
<td>None</td>
</tr>
<tr>
<td>filetype</td>
<td>None</td>
<td>Used only when a Genbank or a GFF file is provided; one of "genbank"<br>or "gff" to be used. Default None infers from file extension.</td>
<td>None</td>
</tr>
</tbody>
</table>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">record_class</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Create a new GraphicRecord from a BioPython Record object.</span>

<span class="ss">        Parameters</span>

<span class="ss">        ----------</span>

<span class="ss">        record</span>

<span class="ss">          A BioPython Record object or the path to a Genbank or a GFF file.</span>

<span class="ss">        record_class</span>

<span class="ss">          The graphic record class to use, e.g. GraphicRecord (default) or</span>

<span class="ss">          CircularGraphicRecord. Strings &#39;circular&#39; and &#39;linear&#39; can also be</span>

<span class="ss">          provided.</span>

<span class="ss">        filetype</span>

<span class="ss">          Used only when a Genbank or a GFF file is provided; one of &quot;</span><span class="n">genbank</span><span class="ss">&quot;</span>

<span class="ss">          or &quot;</span><span class="n">gff</span><span class="ss">&quot; to be used. Default None infers from file extension.</span>

<span class="ss">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">classes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">{</span>

<span class="w">            </span><span class="ss">&quot;linear&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span>

<span class="w">            </span><span class="ss">&quot;circular&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">CircularGraphicRecord</span><span class="p">,</span>

<span class="w">            </span><span class="k">None</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span>

<span class="w">        </span><span class="err">}</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">record_class</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">classes</span><span class="p">:</span>

<span class="w">            </span><span class="n">record_class</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">classes</span><span class="o">[</span><span class="n">record_class</span><span class="o">]</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="nf">str</span><span class="p">)</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="n">hasattr</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;read&quot;</span><span class="p">)</span><span class="err">:</span>

<span class="w">            </span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">load_record</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="n">filetype</span><span class="p">)</span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">features</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">record_class</span><span class="p">(</span>

<span class="w">            </span><span class="n">sequence_length</span><span class="o">=</span><span class="nf">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span>

<span class="w">            </span><span class="k">sequence</span><span class="o">=</span><span class="nf">str</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">.</span><span class="n">defined</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">None</span><span class="p">,</span>

<span class="w">            </span><span class="n">features</span><span class="o">=[</span>

<span class="n">                self.translate_feature(feature)</span>

<span class="n">                for feature in filtered_features</span>

<span class="n">                if feature.location is not None</span>

<span class="n">            </span><span class="o">]</span><span class="p">,</span>

<span class="w">            </span><span class="o">**</span><span class="n">self</span><span class="p">.</span><span class="n">graphic_record_parameters</span>

<span class="w">        </span><span class="p">)</span>
</code></pre></div>

</details>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        
<footer class="md-footer">
    
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../../.." class="md-footer__link md-footer__link--prev" aria-label="Previous: Home" rel="prev">
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <div class="md-ellipsis">
                <span class="md-footer__direction">
                  Previous
                </span>
                Home
              </div>
            </div>
          </a>
        
        
          
          <a href="../Methyl/" class="md-footer__link md-footer__link--next" aria-label="Next: Methyl" rel="next">
            <div class="md-footer__title">
              <div class="md-ellipsis">
                <span class="md-footer__direction">
                  Next
                </span>
                Methyl
              </div>
            </div>
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
            
            Powered by
            <a href="http://timothycrosley.github.io/portray">portray.</a>
            You too can
            <a href="http://timothycrosley.github.io/portray">
              portray</a>
            your Python project well using automatic documentation.
          </div>
        
      </div>
    </div>
  </footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.078830c0.min.js"></script>
      
    
  </body>
</html>