
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.3.1, mkdocs-material-8.5.4">
    
    
      
        <title>Methyl - EpiJinn</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.80dcb947.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.cbb835fc.min.css">
        
          
          
          <meta name="theme-color" content="#4cae4f">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="green" data-md-color-accent="lightgreen">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#module-epijinnmethyl" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="EpiJinn" class="md-header__button md-logo" aria-label="EpiJinn" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            EpiJinn
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Methyl
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    EpiJinn
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="EpiJinn" class="md-nav__button md-logo" aria-label="EpiJinn" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    EpiJinn
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    EpiJinn
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2_1" type="checkbox" id="__nav_2_1" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2_1">
          Epijinn
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Epijinn" data-md-level="2">
        <label class="md-nav__title" for="__nav_2_1">
          <span class="md-nav__icon md-icon"></span>
          Epijinn
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Bedmethyl/" class="md-nav__link">
        Bedmethyl
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Methyl
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Methyl
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functions" class="md-nav__link">
    Functions
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotate_methylation" class="md-nav__link">
    annotate_methylation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methylase" class="md-nav__link">
    Methylase
  </a>
  
    <nav class="md-nav" aria-label="Methylase">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#descendants" class="md-nav__link">
    Descendants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#complement" class="md-nav__link">
    complement
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reverse" class="md-nav__link">
    reverse
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reverse_complement" class="md-nav__link">
    reverse_complement
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methylator" class="md-nav__link">
    Methylator
  </a>
  
    <nav class="md-nav" aria-label="Methylator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extend_restriction_regions" class="md-nav__link">
    extend_restriction_regions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_methylation_sites_in_pattern" class="md-nav__link">
    find_methylation_sites_in_pattern
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_one_methylation_site_in_pattern" class="md-nav__link">
    find_one_methylation_site_in_pattern
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../dnd/" class="md-nav__link">
        Dnd
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reports/" class="md-nav__link">
        Reports
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../version/" class="md-nav__link">
        Version
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functions" class="md-nav__link">
    Functions
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotate_methylation" class="md-nav__link">
    annotate_methylation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methylase" class="md-nav__link">
    Methylase
  </a>
  
    <nav class="md-nav" aria-label="Methylase">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#descendants" class="md-nav__link">
    Descendants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#complement" class="md-nav__link">
    complement
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reverse" class="md-nav__link">
    reverse
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reverse_complement" class="md-nav__link">
    reverse_complement
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methylator" class="md-nav__link">
    Methylator
  </a>
  
    <nav class="md-nav" aria-label="Methylator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extend_restriction_regions" class="md-nav__link">
    extend_restriction_regions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_methylation_sites_in_pattern" class="md-nav__link">
    find_methylation_sites_in_pattern
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_one_methylation_site_in_pattern" class="md-nav__link">
    find_one_methylation_site_in_pattern
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/Edinburgh-Genome-Foundry/EpiJinn/edit/main/reference/epijinn/Methyl.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


<h1 id="module-epijinnmethyl">Module epijinn.Methyl</h1>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="c1"># Copyright 2024 Edinburgh Genome Foundry, University of Edinburgh</span>

<span class="c1">#</span>

<span class="c1"># This file is part of EpiJinn.</span>

<span class="c1">#</span>

<span class="c1"># EpiJinn is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.</span>

<span class="c1">#</span>

<span class="c1"># EpiJinn is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</span>

<span class="c1">#</span>

<span class="c1"># You should have received a copy of the GNU General Public License along with Ediacara. If not, see &lt;https://www.gnu.org/licenses/&gt;.</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span><span class="w"> </span><span class="n">Seq</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqFeature</span><span class="w"> </span><span class="kn">import</span><span class="w"> </span><span class="n">SeqFeature</span><span class="p">,</span><span class="w"> </span><span class="n">FeatureLocation</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">dnachisel</span>

<span class="k">class</span><span class="w"> </span><span class="nc">Methylase</span><span class="p">:</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Methylase enzyme class.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **name**</span>

<span class="s2">    &gt; Name of the enzyme (`str`).</span>

<span class="s2">    **sequence**</span>

<span class="s2">    &gt; Sequence of extended nucleotide characters (`str`).</span>

<span class="s2">    **index_pos**</span>

<span class="s2">    &gt; Index of the methylated base on the positive strand (`int`).</span>

<span class="s2">    **index_neg**</span>

<span class="s2">    &gt; Index of the methylated base on the negative strand (`int`).</span>

<span class="s2">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="n">complement_table</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>

<span class="w">        </span><span class="s2">&quot;A&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;T&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;G&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;C&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;C&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;G&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;T&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;A&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;Y&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;R&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;R&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;Y&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;W&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;W&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;S&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;S&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;K&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;M&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;M&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;K&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;D&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;H&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;V&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;B&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;H&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;D&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;B&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;V&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;X&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;X&quot;</span><span class="p">,</span>

<span class="w">        </span><span class="s2">&quot;N&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;N&quot;</span><span class="p">,</span>

<span class="w">    </span><span class="p">}</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="p">):</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sequence</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">index_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">index_pos</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">index_neg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">index_neg</span>

<span class="w">    </span><span class="nd">@staticmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">reverse</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>

<span class="w">        </span><span class="n">reverse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">reverse</span>

<span class="w">    </span><span class="nd">@staticmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>

<span class="w">        </span><span class="n">complement_letters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="n">Methylase</span><span class="o">.</span><span class="n">complement_table</span><span class="p">[</span><span class="n">letter</span><span class="p">]</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">letter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">sequence</span><span class="p">]</span>

<span class="w">        </span><span class="n">complement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">complement_letters</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">complement</span>

<span class="w">    </span><span class="nd">@staticmethod</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>

<span class="w">        </span><span class="n">r</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="o">.</span><span class="n">reverse</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="o">.</span><span class="n">complement</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">rc</span>

<span class="k">class</span><span class="w"> </span><span class="nc">Methylator</span><span class="p">:</span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;Class for finding methylation sites within a pattern (site) in a sequence.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **sequence**</span>

<span class="s2">    &gt; Sequence of ATGC characters (`str`).</span>

<span class="s2">    **site**</span>

<span class="s2">    &gt; Sequence of restriction enzyme recognition site (`str`).</span>

<span class="s2">    **methylases**</span>

<span class="s2">    &gt; Methylase class instances (`list`). Default uses built-in methylases.</span>

<span class="s2">    &quot;&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="p">,</span><span class="w"> </span><span class="n">site</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sequence</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">methylases</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="kc">None</span><span class="p">:</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">methylases</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">list</span><span class="p">(</span><span class="n">METHYLASES</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<span class="w">        </span><span class="k">else</span><span class="p">:</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">methylases</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylases</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">site</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">site</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">regions_seq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">pattern</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">site_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">pattern_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">site_rc</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">regions_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">pattern_rc</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">regions_seq</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">regions_rc</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">find_methylation_sites_in_pattern</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Run find_one_methylation_site_in_pattern() for each enzyme in methylases.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Matches against methylase enzyme sites:</span><span class="se">\n\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">methylase</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">methylases</span><span class="p">:</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">find_one_methylation_site_in_pattern</span><span class="p">(</span><span class="n">methylase</span><span class="p">)</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">find_one_methylation_site_in_pattern</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Find overlapping methylation and restriction sites.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">extend_restriction_regions</span><span class="p">(</span><span class="n">methylase</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># For matching against positive strand of methylation pattern:</span>

<span class="w">        </span><span class="n">expression</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">DnaNotationPattern</span><span class="o">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># For matching against negative strand of methylation pattern:</span>

<span class="w">        </span><span class="n">expression_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">DnaNotationPattern</span><span class="o">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="o">.</span><span class="n">rc</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression_rc</span><span class="p">)</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;=&quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">extended_regions</span><span class="p">:</span>

<span class="w">            </span><span class="n">region_sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">]</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Region:&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span><span class="n">region</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="n">match_location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">                </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Match in positive strand: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="p">:</span>

<span class="w">                </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Positive strand: -</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="n">match_location_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern_rc</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">match_location_rc</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">                </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Match in negative strand: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="p">:</span>

<span class="w">                </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Negative strand: -</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="bp">self</span><span class="o">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="k">def</span><span class="w"> </span><span class="nf">extend_restriction_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">):</span>

<span class="w">        </span><span class="s2">&quot;&quot;&quot;Modify list of dnachisel.Location of restriction sites to include</span>

<span class="s2">        flanking nucleotides around restriction sites.</span>

<span class="s2">        &quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">regions</span><span class="p">:</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="o">.</span><span class="n">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_neg</span><span class="p">,</span><span class="w"> </span><span class="n">left</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="o">=</span><span class="kc">False</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="c1"># extension upstream</span>

<span class="w">            </span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">)</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="o">.</span><span class="n">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">upper_limit</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">),</span><span class="w"> </span><span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="o">=</span><span class="kc">True</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="c1"># extension downstream</span>

<span class="w">            </span><span class="n">extended_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">extended_regions</span>

<span class="k">def</span><span class="w"> </span><span class="nf">annotate_methylation</span><span class="p">(</span><span class="n">seqrecord</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="n">methylases</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="kc">None</span><span class="p">:</span>

<span class="w">        </span><span class="n">methylases</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">list</span><span class="p">(</span><span class="n">METHYLASES</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">methylase</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">methylases</span><span class="p">:</span>

<span class="w">        </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span>

<span class="w">        </span><span class="n">regex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">DnaNotationPattern</span><span class="o">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">regex</span><span class="p">)</span>

<span class="w">        </span><span class="n">match_location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">seqrecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="n">match</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">match_location</span><span class="p">:</span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;)&quot;</span>

<span class="w">                </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                    </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">end</span><span class="p">),</span><span class="w">  </span><span class="c1"># dnachisel.Location</span>

<span class="w">                        </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="c1"># Mark the methylation site for checking overlap with restriction site</span>

<span class="w">                </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span>

<span class="w">                </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span><span class="p">])</span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=1)&quot;</span>

<span class="w">                </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                    </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="n">FeatureLocation</span><span class="p">(</span>

<span class="w">                            </span><span class="n">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">strand</span><span class="o">=</span><span class="mi">1</span>

<span class="w">                        </span><span class="p">),</span>

<span class="w">                        </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="c1"># Negative strand:</span>

<span class="w">                </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_neg</span>

<span class="w">                </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span>

<span class="w">                    </span><span class="n">Seq</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">methylase</span><span class="o">.</span><span class="n">index_neg</span><span class="p">])</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=-1)&quot;</span>

<span class="w">                </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                    </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="n">FeatureLocation</span><span class="p">(</span>

<span class="w">                            </span><span class="n">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">strand</span><span class="o">=-</span><span class="mi">1</span>

<span class="w">                        </span><span class="p">),</span>

<span class="w">                        </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">        </span><span class="c1"># Repeat for reverse complement, if not palindromic:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">Seq</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">Seq</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">():</span>

<span class="w">            </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span>

<span class="w">            </span><span class="n">methylase_rc_seq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">            </span><span class="n">regex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">DnaNotationPattern</span><span class="o">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase_rc_seq</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="o">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">regex</span><span class="p">)</span>

<span class="w">            </span><span class="n">match_location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">seqrecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">                </span><span class="k">for</span><span class="w"> </span><span class="n">match</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">match_location</span><span class="p">:</span>

<span class="w">                    </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn_rc(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;)&quot;</span>

<span class="w">                    </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                        </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="n">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">end</span>

<span class="w">                            </span><span class="p">),</span><span class="w">  </span><span class="c1"># dnachisel.Location</span>

<span class="w">                            </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn_rc&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="c1"># Mark the methylation site:</span>

<span class="w">                    </span><span class="c1"># reverse complement so need to count backwards, and strand=-1</span>

<span class="w">                    </span><span class="c1"># subtract 1 to account for range</span>

<span class="w">                    </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span>

<span class="w">                    </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">methylase</span><span class="o">.</span><span class="n">index_pos</span><span class="p">])</span>

<span class="w">                    </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=-1)&quot;</span>

<span class="w">                    </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                        </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="n">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="n">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">strand</span><span class="o">=-</span><span class="mi">1</span>

<span class="w">                            </span><span class="p">),</span>

<span class="w">                            </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="c1"># Mark methylation in antisense of the enzyme site:</span>

<span class="w">                    </span><span class="c1"># reverse complement so need to count backwards, and strand=1</span>

<span class="w">                    </span><span class="c1"># subtract 1 to account for range</span>

<span class="w">                    </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">match</span><span class="o">.</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">methylase</span><span class="o">.</span><span class="n">index_neg</span>

<span class="w">                    </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span>

<span class="w">                        </span><span class="n">Seq</span><span class="p">(</span>

<span class="w">                            </span><span class="n">methylase</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">methylase</span><span class="o">.</span><span class="n">index_neg</span><span class="p">]</span>

<span class="w">                        </span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;, strand=1)&quot;</span>

<span class="w">                    </span><span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>

<span class="w">                        </span><span class="n">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="n">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="n">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="n">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">strand</span><span class="o">=</span><span class="mi">1</span>

<span class="w">                            </span><span class="p">),</span>

<span class="w">                            </span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="n">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="n">seqrecord</span>

<span class="c1"># Cytosine:</span>

<span class="n">AluI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;AluI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;AGCT&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">BamHI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;BamHI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">CpG</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;CpG&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;CG&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">EcoKDcm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EcoKDcm&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;CCWGG&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="n">GpC</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GpC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">HaeIII</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;HaeIII&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GGCC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">Hhal</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Hhal&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GCGC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">HpaII</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;HpaII&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;CCGG&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">MspI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;MspI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;CCGG&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Adenine:</span>

<span class="n">EcoBI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EcoBI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;TGANNNNNNNNTGCT&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>

<span class="c1"># EcoGII = Methylase(name=&quot;EcoGII&quot;, sequence=&quot;A&quot;, index_pos=0, index_neg=)  # no valid index_neg value</span>

<span class="n">EcoKDam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EcoKDam&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GATC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">EcoKI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EcoKI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;AACNNNNNNGTGC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="n">EcoRI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EcoRI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;GAATTC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="n">TaqI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;TaqI&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;TCGA&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">index_pos</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="w"> </span><span class="n">index_neg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">METHYLASES</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>

<span class="w">    </span><span class="n">AluI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">AluI</span><span class="p">,</span>

<span class="w">    </span><span class="n">BamHI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">BamHI</span><span class="p">,</span>

<span class="w">    </span><span class="n">CpG</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">CpG</span><span class="p">,</span>

<span class="w">    </span><span class="n">EcoKDcm</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">EcoKDcm</span><span class="p">,</span>

<span class="w">    </span><span class="n">GpC</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">GpC</span><span class="p">,</span>

<span class="w">    </span><span class="n">HaeIII</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">HaeIII</span><span class="p">,</span>

<span class="w">    </span><span class="n">Hhal</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">Hhal</span><span class="p">,</span>

<span class="w">    </span><span class="n">HpaII</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">HpaII</span><span class="p">,</span>

<span class="w">    </span><span class="n">MspI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">MspI</span><span class="p">,</span>

<span class="w">    </span><span class="n">EcoBI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">EcoBI</span><span class="p">,</span>

<span class="w">    </span><span class="c1"># EcoGII.name: EcoGII,</span>

<span class="w">    </span><span class="n">EcoKDam</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">EcoKDam</span><span class="p">,</span>

<span class="w">    </span><span class="n">EcoKI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">EcoKI</span><span class="p">,</span>

<span class="w">    </span><span class="n">EcoRI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">EcoRI</span><span class="p">,</span>

<span class="w">    </span><span class="n">TaqI</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="w"> </span><span class="n">TaqI</span><span class="p">,</span>

<span class="p">}</span>
</code></pre></div>

</details>
<h2 id="variables">Variables</h2>
<div class="codehilite"><pre><span></span><code><span class="n">AluI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">BamHI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">CpG</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">EcoBI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">EcoKDam</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">EcoKDcm</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">EcoKI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">EcoRI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">GpC</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">HaeIII</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">Hhal</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">HpaII</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">METHYLASES</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">MspI</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">TaqI</span>
</code></pre></div>

<h2 id="functions">Functions</h2>
<h3 id="annotate_methylation">annotate_methylation</h3>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_methylation</span><span class="p">(</span>
    <span class="n">seqrecord</span><span class="p">,</span>
    <span class="n">methylases</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="nx">def</span><span class="w"> </span><span class="nx">annotate_methylation</span><span class="p">(</span><span class="nx">seqrecord</span><span class="p">,</span><span class="w"> </span><span class="nx">methylases</span><span class="p">=</span><span class="nx">None</span><span class="p">):</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="nx">methylases</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="nx">None</span><span class="p">:</span>

<span class="w">        </span><span class="nx">methylases</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">list</span><span class="p">(</span><span class="nx">METHYLASES</span><span class="p">.</span><span class="nx">values</span><span class="p">())</span>

<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="nx">methylase</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">methylases</span><span class="p">:</span>

<span class="w">        </span><span class="nx">name</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">name</span>

<span class="w">        </span><span class="nx">regex</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">DnaNotationPattern</span><span class="p">.</span><span class="nx">dna_sequence_to_regexpr</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="nx">pattern</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">SequencePattern</span><span class="p">(</span><span class="nx">regex</span><span class="p">)</span>

<span class="w">        </span><span class="nx">match_location</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">pattern</span><span class="p">.</span><span class="nx">find_matches</span><span class="p">(</span><span class="nx">str</span><span class="p">(</span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">seq</span><span class="p">))</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nx">len</span><span class="p">(</span><span class="nx">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="k">match</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">match_location</span><span class="p">:</span>

<span class="w">                </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;)&quot;</span>

<span class="w">                </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                    </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">FeatureLocation</span><span class="p">(</span><span class="k">match</span><span class="p">.</span><span class="nx">start</span><span class="p">,</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">end</span><span class="p">),</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">Location</span>

<span class="w">                        </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="err">#</span><span class="w"> </span><span class="nx">Mark</span><span class="w"> </span><span class="nx">the</span><span class="w"> </span><span class="nx">methylation</span><span class="w"> </span><span class="nx">site</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">checking</span><span class="w"> </span><span class="nx">overlap</span><span class="w"> </span><span class="nx">with</span><span class="w"> </span><span class="nx">restriction</span><span class="w"> </span><span class="nx">site</span>

<span class="w">                </span><span class="nx">methylated_position</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">start</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_pos</span>

<span class="w">                </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">str</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">[</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_pos</span><span class="p">])</span>

<span class="w">                </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;, strand=1)&quot;</span>

<span class="w">                </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                    </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">FeatureLocation</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="nx">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="mi">1</span>

<span class="w">                        </span><span class="p">),</span>

<span class="w">                        </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="err">#</span><span class="w"> </span><span class="nx">Negative</span><span class="w"> </span><span class="nx">strand</span><span class="p">:</span>

<span class="w">                </span><span class="nx">methylated_position</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">start</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_neg</span>

<span class="w">                </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">str</span><span class="p">(</span>

<span class="w">                    </span><span class="nx">Seq</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">[</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_neg</span><span class="p">]).</span><span class="nx">reverse_complement</span><span class="p">()</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">                </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;, strand=-1)&quot;</span>

<span class="w">                </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                    </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">FeatureLocation</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="nx">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="o">-</span><span class="mi">1</span>

<span class="w">                        </span><span class="p">),</span>

<span class="w">                        </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                        </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                </span><span class="p">)</span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="nx">Repeat</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">reverse</span><span class="w"> </span><span class="nx">complement</span><span class="p">,</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="k">not</span><span class="w"> </span><span class="nx">palindromic</span><span class="p">:</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="nx">Seq</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="nx">Seq</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">).</span><span class="nx">reverse_complement</span><span class="p">():</span>

<span class="w">            </span><span class="nx">name</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">name</span>

<span class="w">            </span><span class="nx">methylase_rc_seq</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">reverse_complement</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">)</span>

<span class="w">            </span><span class="nx">regex</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">DnaNotationPattern</span><span class="p">.</span><span class="nx">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">                </span><span class="nx">methylase_rc_seq</span>

<span class="w">            </span><span class="p">)</span>

<span class="w">            </span><span class="nx">pattern</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">SequencePattern</span><span class="p">(</span><span class="nx">regex</span><span class="p">)</span>

<span class="w">            </span><span class="nx">match_location</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">pattern</span><span class="p">.</span><span class="nx">find_matches</span><span class="p">(</span><span class="nx">str</span><span class="p">(</span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">seq</span><span class="p">))</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="nx">len</span><span class="p">(</span><span class="nx">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="p">:</span>

<span class="w">                </span><span class="k">for</span><span class="w"> </span><span class="k">match</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">match_location</span><span class="p">:</span>

<span class="w">                    </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn_rc(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;)&quot;</span>

<span class="w">                    </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="k">match</span><span class="p">.</span><span class="nx">start</span><span class="p">,</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">end</span>

<span class="w">                            </span><span class="p">),</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="nx">dnachisel</span><span class="p">.</span><span class="nx">Location</span>

<span class="w">                            </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn_rc&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">Mark</span><span class="w"> </span><span class="nx">the</span><span class="w"> </span><span class="nx">methylation</span><span class="w"> </span><span class="nx">site</span><span class="p">:</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">reverse</span><span class="w"> </span><span class="nx">complement</span><span class="w"> </span><span class="nx">so</span><span class="w"> </span><span class="nx">need</span><span class="w"> </span><span class="nx">to</span><span class="w"> </span><span class="nx">count</span><span class="w"> </span><span class="nx">backwards</span><span class="p">,</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="o">-</span><span class="mi">1</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">subtract</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="nx">to</span><span class="w"> </span><span class="nx">account</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">range</span>

<span class="w">                    </span><span class="nx">methylated_position</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_pos</span>

<span class="w">                    </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">str</span><span class="p">(</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">[</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_pos</span><span class="p">])</span>

<span class="w">                    </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;, strand=-1)&quot;</span>

<span class="w">                    </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="nx">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="nx">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="o">-</span><span class="mi">1</span>

<span class="w">                            </span><span class="p">),</span>

<span class="w">                            </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">Mark</span><span class="w"> </span><span class="nx">methylation</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nx">antisense</span><span class="w"> </span><span class="nx">of</span><span class="w"> </span><span class="nx">the</span><span class="w"> </span><span class="nx">enzyme</span><span class="w"> </span><span class="nx">site</span><span class="p">:</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">reverse</span><span class="w"> </span><span class="nx">complement</span><span class="w"> </span><span class="nx">so</span><span class="w"> </span><span class="nx">need</span><span class="w"> </span><span class="nx">to</span><span class="w"> </span><span class="nx">count</span><span class="w"> </span><span class="nx">backwards</span><span class="p">,</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="mi">1</span>

<span class="w">                    </span><span class="err">#</span><span class="w"> </span><span class="nx">subtract</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="nx">to</span><span class="w"> </span><span class="nx">account</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nx">range</span>

<span class="w">                    </span><span class="nx">methylated_position</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="k">match</span><span class="p">.</span><span class="nx">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_neg</span>

<span class="w">                    </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nx">str</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">Seq</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">methylase</span><span class="p">.</span><span class="nx">sequence</span><span class="p">[</span><span class="nx">methylase</span><span class="p">.</span><span class="nx">index_neg</span><span class="p">]</span>

<span class="w">                        </span><span class="p">).</span><span class="nx">reverse_complement</span><span class="p">()</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">                    </span><span class="nx">label</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;@epijinn(&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nx">methylated_nucleotide</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;, strand=1)&quot;</span>

<span class="w">                    </span><span class="nx">seqrecord</span><span class="p">.</span><span class="nx">features</span><span class="p">.</span><span class="nx">append</span><span class="p">(</span>

<span class="w">                        </span><span class="nx">SeqFeature</span><span class="p">(</span>

<span class="w">                            </span><span class="nx">FeatureLocation</span><span class="p">(</span>

<span class="w">                                </span><span class="nx">methylated_position</span><span class="p">,</span><span class="w"> </span><span class="nx">methylated_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="nx">strand</span><span class="p">=</span><span class="mi">1</span>

<span class="w">                            </span><span class="p">),</span>

<span class="w">                            </span><span class="k">type</span><span class="p">=</span><span class="s">&quot;misc_feature&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">id</span><span class="p">=</span><span class="s">&quot;@epijinn&quot;</span><span class="p">,</span>

<span class="w">                            </span><span class="nx">qualifiers</span><span class="p">={</span><span class="s">&quot;label&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">label</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;note&quot;</span><span class="p">:</span><span class="w"> </span><span class="nx">name</span><span class="p">},</span>

<span class="w">                        </span><span class="p">)</span>

<span class="w">                    </span><span class="p">)</span>

<span class="w">    </span><span class="k">return</span><span class="w"> </span><span class="nx">seqrecord</span>
</code></pre></div>

</details>
<h2 id="classes">Classes</h2>
<h3 id="methylase">Methylase</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">Methylase</span><span class="p">(</span>
    <span class="n">name</span><span class="p">,</span>
    <span class="n">sequence</span><span class="p">,</span>
    <span class="n">index_pos</span><span class="p">,</span>
    <span class="n">index_neg</span>
<span class="p">)</span>
</code></pre></div>

<p>Methylase enzyme class.</p>
<p><strong>Parameters</strong></p>
<p><strong>name</strong></p>
<blockquote>
<p>Name of the enzyme (<code>str</code>).</p>
</blockquote>
<p><strong>sequence</strong></p>
<blockquote>
<p>Sequence of extended nucleotide characters (<code>str</code>).</p>
</blockquote>
<p><strong>index_pos</strong></p>
<blockquote>
<p>Index of the methylated base on the positive strand (<code>int</code>).</p>
</blockquote>
<p><strong>index_neg</strong></p>
<blockquote>
<p>Index of the methylated base on the negative strand (<code>int</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="s s-Atom">class</span> <span class="nv">Methylase</span><span class="o">:</span>

    <span class="s2">&quot;&quot;&quot;Methylase enzyme class.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **name**</span>

<span class="s2">    &gt; Name of the enzyme (`str`).</span>

<span class="s2">    **sequence**</span>

<span class="s2">    &gt; Sequence of extended nucleotide characters (`str`).</span>

<span class="s2">    **index_pos**</span>

<span class="s2">    &gt; Index of the methylated base on the positive strand (`int`).</span>

<span class="s2">    **index_neg**</span>

<span class="s2">    &gt; Index of the methylated base on the negative strand (`int`).</span>

<span class="s2">    &quot;&quot;&quot;</span>

    <span class="s s-Atom">complement_table</span> <span class="o">=</span> <span class="p">{</span>

        <span class="s2">&quot;A&quot;</span><span class="o">:</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span>

        <span class="s2">&quot;G&quot;</span><span class="o">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span>

        <span class="s2">&quot;C&quot;</span><span class="o">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>

        <span class="s2">&quot;T&quot;</span><span class="o">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>

        <span class="s2">&quot;Y&quot;</span><span class="o">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>

        <span class="s2">&quot;R&quot;</span><span class="o">:</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span>

        <span class="s2">&quot;W&quot;</span><span class="o">:</span> <span class="s2">&quot;W&quot;</span><span class="p">,</span>

        <span class="s2">&quot;S&quot;</span><span class="o">:</span> <span class="s2">&quot;S&quot;</span><span class="p">,</span>

        <span class="s2">&quot;K&quot;</span><span class="o">:</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>

        <span class="s2">&quot;M&quot;</span><span class="o">:</span> <span class="s2">&quot;K&quot;</span><span class="p">,</span>

        <span class="s2">&quot;D&quot;</span><span class="o">:</span> <span class="s2">&quot;H&quot;</span><span class="p">,</span>

        <span class="s2">&quot;V&quot;</span><span class="o">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span>

        <span class="s2">&quot;H&quot;</span><span class="o">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span>

        <span class="s2">&quot;B&quot;</span><span class="o">:</span> <span class="s2">&quot;V&quot;</span><span class="p">,</span>

        <span class="s2">&quot;X&quot;</span><span class="o">:</span> <span class="s2">&quot;X&quot;</span><span class="p">,</span>

        <span class="s2">&quot;N&quot;</span><span class="o">:</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span>

    <span class="p">}</span>

    <span class="s s-Atom">def</span> <span class="k">__</span><span class="nf">init__</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">,</span> <span class="s s-Atom">name</span><span class="p">,</span> <span class="s s-Atom">sequence</span><span class="p">,</span> <span class="s s-Atom">index_pos</span><span class="p">,</span> <span class="s s-Atom">index_neg</span><span class="p">)</span><span class="o">:</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">name</span> <span class="o">=</span> <span class="s s-Atom">name</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">sequence</span> <span class="o">=</span> <span class="s s-Atom">sequence</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">rc</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nf">reverse_complement</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">index_pos</span> <span class="o">=</span> <span class="s s-Atom">index_pos</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">index_neg</span> <span class="o">=</span> <span class="s s-Atom">index_neg</span>

    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">reverse</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span><span class="o">:</span>

        <span class="s s-Atom">reverse</span> <span class="o">=</span> <span class="s s-Atom">sequence</span><span class="p">[</span><span class="o">::-</span><span class="mi">1</span><span class="p">]</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">reverse</span>

    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">complement</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span><span class="o">:</span>

        <span class="s s-Atom">complement_letters</span> <span class="o">=</span> <span class="p">[</span><span class="nv">Methylase</span><span class="p">.</span><span class="s s-Atom">complement_table</span><span class="p">[</span><span class="s s-Atom">letter</span><span class="p">]</span> <span class="s s-Atom">for</span> <span class="s s-Atom">letter</span> <span class="s s-Atom">in</span> <span class="s s-Atom">sequence</span><span class="p">]</span>

        <span class="s s-Atom">complement</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">.</span><span class="nf">join</span><span class="p">(</span><span class="s s-Atom">complement_letters</span><span class="p">)</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">complement</span>

    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span><span class="o">:</span>

        <span class="s s-Atom">r</span> <span class="o">=</span> <span class="nv">Methylase</span><span class="p">.</span><span class="nf">reverse</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span>

        <span class="s s-Atom">rc</span> <span class="o">=</span> <span class="nv">Methylase</span><span class="p">.</span><span class="nf">complement</span><span class="p">(</span><span class="s s-Atom">r</span><span class="p">)</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">rc</span>
</code></pre></div>

</details>
<hr />
<h4 id="descendants">Descendants</h4>
<ul>
<li>epijinn.dnd.Dnd</li>
</ul>
<h4 id="class-variables">Class variables</h4>
<div class="codehilite"><pre><span></span><code><span class="n">complement_table</span>
</code></pre></div>

<h4 id="static-methods">Static methods</h4>
<h4 id="complement">complement</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span>
    <span class="n">sequence</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="nv">@staticmethod</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">complement</span><span class="p">(</span><span class="k">sequence</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">complement_letters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span><span class="n">Methylase.complement_table[letter</span><span class="o">]</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">letter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="k">sequence</span><span class="err">]</span>

<span class="w">        </span><span class="n">complement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">complement_letters</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">complement</span>
</code></pre></div>

</details>
<h4 id="reverse">reverse</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">reverse</span><span class="p">(</span>
    <span class="n">sequence</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">reverse</span><span class="p">(</span><span class="s s-Atom">sequence</span><span class="p">)</span><span class="o">:</span>

        <span class="s s-Atom">reverse</span> <span class="o">=</span> <span class="s s-Atom">sequence</span><span class="p">[</span><span class="o">::-</span><span class="mi">1</span><span class="p">]</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">reverse</span>
</code></pre></div>

</details>
<h4 id="reverse_complement">reverse_complement</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span>
    <span class="n">sequence</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="nv">@staticmethod</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">reverse_complement</span><span class="p">(</span><span class="k">sequence</span><span class="p">)</span><span class="err">:</span>

<span class="w">        </span><span class="n">r</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">.</span><span class="nf">reverse</span><span class="p">(</span><span class="k">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">.</span><span class="n">complement</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">rc</span>
</code></pre></div>

</details>
<h3 id="methylator">Methylator</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">Methylator</span><span class="p">(</span>
    <span class="n">sequence</span><span class="p">,</span>
    <span class="n">site</span><span class="p">,</span>
    <span class="n">methylases</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Class for finding methylation sites within a pattern (site) in a sequence.</p>
<p><strong>Parameters</strong></p>
<p><strong>sequence</strong></p>
<blockquote>
<p>Sequence of ATGC characters (<code>str</code>).</p>
</blockquote>
<p><strong>site</strong></p>
<blockquote>
<p>Sequence of restriction enzyme recognition site (<code>str</code>).</p>
</blockquote>
<p><strong>methylases</strong></p>
<blockquote>
<p>Methylase class instances (<code>list</code>). Default uses built-in methylases.</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="n">class</span><span class="w"> </span><span class="n">Methylator</span><span class="o">:</span>

<span class="w">    </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Class for finding methylation sites within a pattern (site) in a sequence.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **sequence**</span>

<span class="s2">    &gt; Sequence of ATGC characters (`str`).</span>

<span class="s2">    **site**</span>

<span class="s2">    &gt; Sequence of restriction enzyme recognition site (`str`).</span>

<span class="s2">    **methylases**</span>

<span class="s2">    &gt; Methylase class instances (`list`). Default uses built-in methylases.</span>

<span class="s2">    </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">__init__</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">sequence</span><span class="p">,</span><span class="w"> </span><span class="n">site</span><span class="p">,</span><span class="w"> </span><span class="n">methylases</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sequence</span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">methylases</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="k">None</span><span class="o">:</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">methylases</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">list</span><span class="p">(</span><span class="n">METHYLASES</span><span class="p">.</span><span class="k">values</span><span class="p">())</span>

<span class="w">        </span><span class="k">else</span><span class="o">:</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">methylases</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">methylases</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">site</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">site</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">regions_seq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">pattern</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">site_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Methylase</span><span class="p">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">pattern_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">site_rc</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">regions_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">pattern_rc</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">regions_seq</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">regions_rc</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">find_methylation_sites_in_pattern</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Run find_one_methylation_site_in_pattern() for each enzyme in methylases.</span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Matches against methylase enzyme sites:</span><span class="se">\n\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">methylase</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">methylases</span><span class="o">:</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">find_one_methylation_site_in_pattern</span><span class="p">(</span><span class="n">methylase</span><span class="p">)</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">find_one_methylation_site_in_pattern</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Find overlapping methylation and restriction sites.</span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">extend_restriction_regions</span><span class="p">(</span><span class="n">methylase</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># For matching against positive strand of methylation pattern:</span>

<span class="w">        </span><span class="n">expression</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">DnaNotationPattern</span><span class="p">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="p">.</span><span class="n">sequence</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># For matching against negative strand of methylation pattern:</span>

<span class="w">        </span><span class="n">expression_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">DnaNotationPattern</span><span class="p">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="p">.</span><span class="n">rc</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression_rc</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="k">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;=&quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="k">name</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">extended_regions</span><span class="o">:</span>

<span class="w">            </span><span class="n">region_sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="err">[</span><span class="n">region</span><span class="p">.</span><span class="k">start</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="k">end</span><span class="err">]</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Region:&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">str</span><span class="p">(</span><span class="n">region</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="n">match_location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Match in positive strand: %s</span><span class="se">\n</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Positive strand: -</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="n">match_location_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern_rc</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">match_location_rc</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mi">0</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Match in negative strand: %s</span><span class="se">\n</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;Negative strand: -</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">extend_restriction_regions</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Modify list of dnachisel.Location of restriction sites to include</span>

<span class="s2">        flanking nucleotides around restriction sites.</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">regions</span><span class="o">:</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="k">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="p">.</span><span class="n">index_neg</span><span class="p">,</span><span class="w"> </span><span class="k">left</span><span class="o">=</span><span class="no">True</span><span class="p">,</span><span class="w"> </span><span class="k">right</span><span class="o">=</span><span class="no">False</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="c1"># extension upstream</span>

<span class="w">            </span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="n">sequence</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="n">index_pos</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">)</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="k">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">upper_limit</span><span class="o">=</span><span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="p">),</span><span class="w"> </span><span class="k">left</span><span class="o">=</span><span class="no">False</span><span class="p">,</span><span class="w"> </span><span class="k">right</span><span class="o">=</span><span class="no">True</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="c1"># extension downstream</span>

<span class="w">            </span><span class="n">extended_regions</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">extended_regions</span>
</code></pre></div>

</details>
<hr />
<h4 id="methods">Methods</h4>
<h4 id="extend_restriction_regions">extend_restriction_regions</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">extend_restriction_regions</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">methylase</span>
<span class="p">)</span>
</code></pre></div>

<p>Modify list of dnachisel.Location of restriction sites to include</p>
<p>flanking nucleotides around restriction sites.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">extend_restriction_regions</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s">&quot;&quot;&quot;Modify list of dnachisel.Location of restriction sites to include</span>

<span class="w">        </span><span class="n">flanking</span><span class="w"> </span><span class="n">nucleotides</span><span class="w"> </span><span class="n">around</span><span class="w"> </span><span class="n">restriction</span><span class="w"> </span><span class="n">sites</span><span class="p">.</span>

<span class="w">        </span><span class="s">&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">regions:</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="n">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">methylase</span><span class="p">.</span><span class="n">index_neg</span><span class="p">,</span><span class="w"> </span><span class="n">left</span><span class="o">=</span><span class="n">True</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="o">=</span><span class="n">False</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="p">#</span><span class="w"> </span><span class="n">extension</span><span class="w"> </span><span class="n">upstream</span>

<span class="w">            </span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="n">sequence</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="n">index_pos</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mh">1</span><span class="p">)</span>

<span class="w">            </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="n">extended</span><span class="p">(</span>

<span class="w">                </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">upper_limit</span><span class="o">=</span><span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="p">),</span><span class="w"> </span><span class="n">left</span><span class="o">=</span><span class="n">False</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="o">=</span><span class="n">True</span>

<span class="w">            </span><span class="p">)</span><span class="w">  </span><span class="p">#</span><span class="w"> </span><span class="n">extension</span><span class="w"> </span><span class="n">downstream</span>

<span class="w">            </span><span class="n">extended_regions</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">extended_regions</span>
</code></pre></div>

</details>
<h4 id="find_methylation_sites_in_pattern">find_methylation_sites_in_pattern</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">find_methylation_sites_in_pattern</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<p>Run find_one_methylation_site_in_pattern() for each enzyme in methylases.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def find_methylation_sites_in_pattern(self):

        &quot;&quot;&quot;Run find_one_methylation_site_in_pattern() for each enzyme in methylases.&quot;&quot;&quot;

        self.report += &quot;Matches against methylase enzyme sites:\n\n&quot;

        for methylase in self.methylases:

            self.find_one_methylation_site_in_pattern(methylase)

            self.report += &quot;\n&quot;
</code></pre></div>

</details>
<h4 id="find_one_methylation_site_in_pattern">find_one_methylation_site_in_pattern</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">find_one_methylation_site_in_pattern</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">methylase</span>
<span class="p">)</span>
</code></pre></div>

<p>Find overlapping methylation and restriction sites.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">find_one_methylation_site_in_pattern</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">methylase</span><span class="p">)</span><span class="o">:</span>

<span class="w">        </span><span class="s">&quot;&quot;&quot;Find overlapping methylation and restriction sites.&quot;&quot;&quot;</span>

<span class="w">        </span><span class="n">extended_regions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">extend_restriction_regions</span><span class="p">(</span><span class="n">methylase</span><span class="p">)</span>

<span class="w">        </span><span class="p">#</span><span class="w"> </span><span class="n">For</span><span class="w"> </span><span class="n">matching</span><span class="w"> </span><span class="n">against</span><span class="w"> </span><span class="n">positive</span><span class="w"> </span><span class="n">strand</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">methylation</span><span class="w"> </span><span class="nl">pattern:</span>

<span class="w">        </span><span class="n">expression</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">DnaNotationPattern</span><span class="p">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="p">.</span><span class="n">sequence</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression</span><span class="p">)</span>

<span class="w">        </span><span class="p">#</span><span class="w"> </span><span class="n">For</span><span class="w"> </span><span class="n">matching</span><span class="w"> </span><span class="n">against</span><span class="w"> </span><span class="n">negative</span><span class="w"> </span><span class="n">strand</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">methylation</span><span class="w"> </span><span class="nl">pattern:</span>

<span class="w">        </span><span class="n">expression_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">DnaNotationPattern</span><span class="p">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span>

<span class="w">            </span><span class="n">methylase</span><span class="p">.</span><span class="n">rc</span>

<span class="w">        </span><span class="p">)</span>

<span class="w">        </span><span class="n">pattern_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dnachisel</span><span class="p">.</span><span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression_rc</span><span class="p">)</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="n">methylase</span><span class="p">.</span><span class="n">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;=&quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">methylase</span><span class="p">.</span><span class="n">name</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nl">extended_regions:</span>

<span class="w">            </span><span class="n">region_sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">region</span><span class="p">.</span><span class="n">start</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">region</span><span class="p">.</span><span class="k">end</span><span class="p">]</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;Region:&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">str</span><span class="p">(</span><span class="n">region</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>

<span class="w">            </span><span class="n">match_location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">match_location</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mh">0</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;Match in positive strand: %s</span><span class="se">\n</span><span class="s">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;Positive strand: -</span><span class="se">\n</span><span class="s">&quot;</span>

<span class="w">            </span><span class="n">match_location_rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pattern_rc</span><span class="p">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">region_sequence</span><span class="p">)</span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">len</span><span class="p">(</span><span class="n">match_location_rc</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="mh">0</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;Match in negative strand: %s</span><span class="se">\n</span><span class="s">&quot;</span><span class="w"> </span><span class="o">%</span><span class="w"> </span><span class="n">region_sequence</span>

<span class="w">            </span><span class="k">else</span><span class="o">:</span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;Negative strand: -</span><span class="se">\n</span><span class="s">&quot;</span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">report</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>
</code></pre></div>

</details>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        
<footer class="md-footer">
    
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../Bedmethyl/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Bedmethyl" rel="prev">
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <div class="md-ellipsis">
                <span class="md-footer__direction">
                  Previous
                </span>
                Bedmethyl
              </div>
            </div>
          </a>
        
        
          
          <a href="../dnd/" class="md-footer__link md-footer__link--next" aria-label="Next: Dnd" rel="next">
            <div class="md-footer__title">
              <div class="md-ellipsis">
                <span class="md-footer__direction">
                  Next
                </span>
                Dnd
              </div>
            </div>
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
            
            Powered by
            <a href="http://timothycrosley.github.io/portray">portray.</a>
            You too can
            <a href="http://timothycrosley.github.io/portray">
              portray</a>
            your Python project well using automatic documentation.
          </div>
        
      </div>
    </div>
  </footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.078830c0.min.js"></script>
      
    
  </body>
</html>